---
title: "Reviewer_analyses"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE)
library(GenomicRanges)
library(rtracklayer)
library(tidyverse)
source("/Shares/rinn_class/data/CLASS_2020/analysis/util/_setup.R")
source("/Shares/rinn_class/data/CLASS_2020/analysis/util/intersect_functions.R")



# source("util/plotting_functions.R")
# source("util/intersect_functions.R")
# source("util/_setup.R")
```

```{r}

consensus_peaks <- create_consensus_peaks(broadpeakfilepath = "/Shares/rinn_class/data/k562_chip/results/bwa/mergedLibrary/macs/broadPeak")
# Filter based on number of peaks to be consistent with the previous data.
consensus_peaks <- consensus_peaks[sapply(consensus_peaks, length) > 250]


res_df <- read_csv("../08_defining_reservoirs/results/08_peak_occurence_df_resvoirs.csv")
all_promoters <-  rtracklayer::import("../01_consensus_peaks/results/lncrna_mrna_promoters.gtf")

```


```{r}
# Profile function
profile_tss <- function(peaks, 
                        promoters_gr,
                        upstream = 3e3,
                        downstream = 3e3) {
  

  peak_coverage <- coverage(peaks)
  
  coverage_length <- elementNROWS(peak_coverage)
  coverage_gr <- GRanges(seqnames = names(coverage_length),
                         IRanges(start = rep(1, length(coverage_length)), 
                                 end = coverage_length))
  
  promoters_gr <- subsetByOverlaps(promoters_gr, 
                                       coverage_gr, 
                                       type="within", 
                                       ignore.strand=TRUE)
  chromosomes <- intersect(names(peak_coverage), 
                           unique(as.character(seqnames(promoters_gr))))
  peak_coverage <- peak_coverage[chromosomes]
  
  promoters_ir <- as(promoters_gr, "IntegerRangesList")[chromosomes]
  
  promoter_peak_view <- Views(peak_coverage, promoters_ir)
  
  promoter_peak_view <- lapply(promoter_peak_view, function(x) t(viewApply(x, as.vector)))
  promoter_peak_matrix <- do.call("rbind", promoter_peak_view)
  
  minus_idx <- which(as.character(strand(promoters_gr)) == "-")
  promoter_peak_matrix[minus_idx,] <- promoter_peak_matrix[minus_idx,
                                                           ncol(promoter_peak_matrix):1]
  
  if(length(which(rowSums(promoter_peak_matrix) > 1)) > 1) {
    promoter_peak_matrix <- promoter_peak_matrix[rowSums(promoter_peak_matrix) > 1,]
    
    peak_sums <- colSums(promoter_peak_matrix)
    peak_dens <- peak_sums/sum(peak_sums)
    
    metaplot_df <- data.frame(x = -upstream:(downstream-1),
                              dens = peak_dens)
  } else if(length(which(rowSums(promoter_peak_matrix) > 1)) == 1) {
    promoter_peak_matrix <- t(as.matrix(promoter_peak_matrix[rowSums(promoter_peak_matrix) > 1,]))
    
    peak_sums <- colSums(promoter_peak_matrix)
    peak_dens <- peak_sums/sum(peak_sums)
    
    metaplot_df <- data.frame(x = -upstream:(downstream-1),
                              dens = peak_dens)
  } else {
    metaplot_df <- data.frame(x = -upstream:(downstream-1),
                              dens = 0)
  }
  
  return(metaplot_df)
}
```

```{r}
reservoirs <- res_df %>% filter(reservoir == 1)
nonres <- res_df %>% filter(reservoir == 0)
reservoir_prom <- all_promoters[all_promoters$gene_id %in% reservoirs$gene_id]
nonres_prom <- all_promoters[all_promoters$gene_id %in% nonres$gene_id]

profile_df <- data.frame("x" = integer(),
                         "dens" = numeric(),
                         "reservoir" = character(),
                         "dbp" = character())

for(i in 1:length(consensus_peaks)) {
  pks <- consensus_peaks[[i]]
  
  length(pks)
  reservoir_prof <- profile_tss(pks, reservoir_prom)
  nonres_prof <- profile_tss(pks, nonres_prom)
  
  # label 
  reservoir_prof$reservoir <- "reservoir"
  nonres_prof$reservoir <- "non-reservoir"
  
  reservoir_prof$dbp <- names(consensus_peaks)[i]
  nonres_prof$dbp <- names(consensus_peaks)[i]
  
  profile_df <- bind_rows(profile_df, reservoir_prof, nonres_prof)

  }

write_csv(profile_df, "results/reservoir_tss_metaplot_profile.csv")


```

```{r}
dbps <- unique(profile_df$dbp)


ggplot(profile_df, 
       aes(x = x, y = dens, color = reservoir)) + 
    geom_vline(xintercept = 0, lty = 2) + 
  geom_line(size = 1.5, alpha = 0.7) + 
  facet_wrap(~dbp) +
  ggtitle("Reservoir vs non-reservoir TSS profiles") + 
  scale_x_continuous(breaks = c(-3000, 0, 3000),
                     labels = c("-3kb", "TSS", "+3kb"),
                     name = "") + 
  ylab("Peak frequency") + 
  theme_paperwhite() + 
  scale_color_manual(values = c("#424242","#A8404C"))
ggsave("figures/tss_profile_reservoir_vs_nonres.pdf", height = 20, width = 20)

```







Hommer genome locations
```{r}

base_path <- paste0("/Shares/rinn_class/data/k562_chip/results/bwa/mergedLibrary/macs/broadPeak/consensus/")

dbps <- unique(gsub("eGFP-", "", names(consensus_peaks)))

annotation_df <- data.frame(annotation = character(),
                            count = numeric(),
                            dbp = character())

for(i in 1:length(dbps)) {
  pks_annotated <- read.table(paste0(base_path, dbps[i], "/", dbps[i], ".consensus_peaks.boolean.annotatePeaks.txt"),
                              sep = "\t", header = T)
  pks_annotated$annot <- sapply(pks_annotated$Annotation, function(x) {
    unlist(strsplit(x, " "))[[1]]
  })
  annotation_df <- bind_rows(annotation_df, 
                             data.frame(annotation = names(table(pks_annotated$annot)),
                                        count = table(pks_annotated$annot),
                                        dbp = dbps[i]))
  
}


annotation_df <- annotation_df %>%
  distinct() %>%
  group_by(dbp) %>%
  mutate(fraction = count.Freq / sum(count.Freq))


write_csv(annotation_df, "results/dbp_genome_location_annotation.csv")
```

```{r}

annotation_df <- read_csv("results/dbp_genome_location_annotation.csv")
# Cluster this
annotation_matrix <- annotation_df %>% 
  dplyr::select(dbp, annotation, fraction) %>%
  pivot_wider(names_from = annotation, values_from = fraction, values_fill = 0) %>%
  column_to_rownames("dbp") %>%
  as.matrix()

annotation_clust <- hclust(dist(annotation_matrix))

# row_clusters <- cutree(annotation_clust, k = 6)

row_ordering <- annotation_clust$labels[annotation_clust$order]
annotation_clust <- hclust(dist(t(annotation_matrix)))
col_ordering <- annotation_clust$labels[annotation_clust$order]

# annotation_df <- merge(annotation_df, data.frame(dbp = names(row_clusters),
#                                                 cluster = row_clusters))


# Set factor orders
annotation_df$dbp <- factor(annotation_df$dbp, levels = row_ordering)
annotation_df$count.Var1 <- factor(annotation_df$count.Var1, levels = col_ordering)
ggplot(annotation_df, aes(x = dbp, y = fraction, fill = count.Var1)) + 
  geom_bar(stat = "identity", position = "stack") + 
  coord_flip() + 
  ggtitle("HOMER peak annotation") 
  # facet_grid(cluster~., scales = "free", space = "free") 
ggsave("figures/peak_annotation_fraction.pdf", height = 15, width = 7)
```





