---
title: "11_reservoir_chromatin_properties"
date: "5/29/2020"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
library(eulerr)
source("../util/intersect_functions.R")
source("../util/_setup.R")

```

#Goal: to determine the chromatin environment of reservoir promoters using 
histone modification ChIP data. Specifically, Histone H3 (H3) Lysine 27 
tri-methylation (H3K27me3) that represents heterochraomtic regions / 
and H3 lysine 27 acetylation that represents euchromatin.

```{r} 

# First we will go get the .bed files of significant peaks determined by encode in k562
# K27me3 - https://www.encodeproject.org/files/ENCFF031FSF/@@download/ENCFF031FSF.bed.gz
# K27ac - https://www.encodeproject.org/files/ENCFF038DDS/@@download/ENCFF038DDS.bed.gz


# Retrieve Histone mark and # TODO add unzip

system("cd data;
           wget https://www.encodeproject.org/files/ENCFF031FSF/@@download/ENCFF031FSF.bed.gz",
           intern = FALSE, ignore.stdout = TRUE, ignore.stderr = TRUE)
system("cd data;
           wget https://www.encodeproject.org/files/ENCFF038DDS/@@download/ENCFF038DDS.bed.gz",
           intern = FALSE, ignore.stdout = TRUE, ignore.stderr = TRUE)

system("cd data;
       gunzip *.gz")


# Organzing the histone mark data into a data frame and then granges of start and stop of peaks in .bed files

peak_files <- data.frame("file" = c(as.character("data/ENCFF031FSF.bed"), as.character("data/ENCFF038DDS.bed")),
                         "histone_mark" = c("H3K27me3", "H3K27ac"))
hmark_peaks <- lapply(as.character(peak_files$file), read.table)
names(hmark_peaks) <- as.character(peak_files$histone_mark)
hmark_granges <- lapply(hmark_peaks, function(peaks) {
  GRanges(seqnames = peaks$V1,
          ranges = IRanges(start = peaks$V2,
                           end = peaks$V3))
         })

# To find overlaps with promoters using gRanges we will open up the GTF of promoters from 01_consensus_peaks
promoters <- rtracklayer::import("../01_consensus_peaks/results/lncrna_mrna_promoters.gtf")

# Read our compiled peak_occurrence data frame from 10_reservoir_nascent_txn (with all the variables for each promoter observation). 
peak_occurrence_df <- read_csv("../10_reservoir_nascent_txn/results/peak_occurrence_df_nascent_added.csv")


# K27me3 Promoter overlaps
ov1 <- findOverlaps(promoters, hmark_granges[[1]])
overlaps <- unique(ov1@from)
overlapping_genes <- promoters$gene_id[overlaps]
peak_occurrence_df$h3k27me3 <- peak_occurrence_df$gene_id %in% overlapping_genes
table(peak_occurrence_df$h3k27me3)

# K27ac Promoter overlaps
ov2 <- findOverlaps(promoters, hmark_granges[[2]])
overlaps <- unique(ov2@from)
overlapping_genes <- promoters$gene_id[overlaps]
peak_occurrence_df$h3k27ac <- peak_occurrence_df$gene_id %in% overlapping_genes
table(peak_occurrence_df$h3k27ac)


# Writting out final peak_occurrence_df !
# First have to change histone mark overlaps to numeric from logical
peak_occurrence_df$k27me3 <- as.numeric(peak_occurrence_df$h3k27me3)
peak_occurrence_df$k27ac <- as.numeric(peak_occurrence_df$h3k27ac)

#remove old histone logical call columns
peak_occurrence_df <- subset(peak_occurrence_df, select = -c(19,20))

# Here it is: the final peak_occurrence_df

write.csv(peak_occurrence_df, "results/peak_occurence_df.csv")

```


# First plotting the density of nascent on reservoir vs non-reservoir
```{r}
# Let's plot nascent transcription relative to the number of DBPs bound.

ggplot(peak_occurrence_df, aes( x = log10(nascentTpm + 1), 
                                colour = as.factor(reservoir))) +
geom_density()
ggsave("figures/reservoir_vs_nascent.pdf")

```


# Figure 5a DBP-vs-Nascent TPM
```{r}

ggplot(peak_occurrence_df, aes(y = log10(nascentTpm + 1), x = number_of_dbp)) +
  geom_smooth()

ggsave("figures/DBPs_vs_nascent.pdf")
```
# Here we found a similar trend as RNA-seq where there is a positive 
correlation between number of DBPs and amount of nascent expression. There is 
also, similar to RNA-seq.


# lets see if there are reservoirs that lack nascent txn (ghost)
# Define Ghost and Zombies and write to peak occurence df
```{r}

ghost <- peak_occurrence_df %>% filter(nascentTpm < 0.001 & reservoir == 1) 

peak_occurrence_df$ghost <- as.numeric(peak_occurrence_df$nascentTpm < 0.001 & 
                                         peak_occurrence_df$reservoir == 1)

ggsave("figures/ghost_vs_nascent.pdf")

```

# Here we find that 355 of 1,362 (26%) of reservoirs have neither mature 
RNA-seq expression nor nascent expression despite many DBPs bound. let's look
at those two distributions, but first define zombies as nascent expression but
not RNA-seq expression from these promoters. 


```{r}
# Adding zombies to peak occurence df.

zombie <- peak_occurrence_df %>% 
  filter(nascentTpm > 0.001 & reservoir == 1) 

peak_occurrence_df$zombie <- as.numeric(peak_occurrence_df$nascentTpm > 0.001 
                                        & peak_occurrence_df$reservoir == 1)

# write out final peak_occurence_df 

write_csv(peak_occurrence_df, "results/peak_occurence_df.csv")

```

# Now we want to explore if the lack of nascent transcripts is due to chromatin
environment. Specifically K27me3 and K27ac states of these promoters.
```{r}

# How it compares to RNA seq

ggplot(peak_occurrence_df, aes(x = log10(tpm + .1), 
                               color = as.logical(k27ac))) +
  geom_density()

#K27ac is positively associated RNA-seq as expected

ggplot(peak_occurrence_df, aes(x = log10(nascentTpm + .1), 
                               color = as.logical(k27ac))) +
  geom_density()

# in general k27ac correlates with higher nascent tpm as expected

# Now narrowing in on only reservoirs looking at K27ac status

ggplot(peak_occurrence_df, aes(x = log10(nascentTpm + 1), 
                               color = as.logical(k27ac & reservoir))) +
  geom_density()

# Suprisingly we do not see a strong enrichment of k27ac towards higher tpm.

# Now let's look at k27me3 status

ggplot(peak_occurrence_df, aes(x = log10(nascentTpm + .1), 
                               color = as.logical(k27me3 & reservoir))) +
  geom_density()

# K27me3 does seperate ghosts from zombies

ggsave("figures/reservoir")


```



# Figure 5b: # DBP distribution of DBPs between ghost and zombie
```{r}

ggplot(peak_occurrence_df %>% filter(reservoir == 1), aes(x = number_of_dbp, 
                               fill = factor(ghost))) +
  geom_density(alpha = .2)

ggplot(peak_occurrence_df %>% filter(reservoir == 1),aes(x = factor(ghost),
                                                         y = number_of_dbp)) +
  geom_boxplot() +
  stat_compare_means()


```




```{r}
#plotting histone occurrence and reservoirr occurrence

genes.venn <- euler(peak_occurrence_df[,c("reservoir","k27me3", "k27ac")])
plot(genes.venn, quantities = TRUE)
ggsave("figures/venn_chroamtin_ov_resvoirs.pdf")


```
---------------
