---
title: "Transposable Elements: Intersect features with peaks"
output: html_document
editor_options:
  chunk_output_type: console
---
```{r setup, include=FALSE}
options(stringsAsFactors = F)
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
library(GenomicFeatures)
library(rtracklayer)
#library(ggbio) unnecessary
library(ggrepel)
```
## Purpose:
When you are finished, this script will intersect either mRNA or lncRNA gene bodies or TE families
with the consensus peak set for each TF. This will result in a bed file for each TF consisting of
the subset of peaks that fall that intersect with that category of genomic feature.
### Supplies:
You will need to gather the Gencode annotation used in the NF-CORE pipeline (v32), the
consensus peak files for each TF, and if you're doing the TE overlap you'll need the repeat masker
table from UCSC.
```{r import}
# Gather gencode as GRanges
#gtf <- rtracklayer::import("gencode.v32.annotation.gtf")
# Gather rmsk.txt file for repeats
rmsk_df <- read.table("/Shares/rinn_class/data/k562_chip/util/rmsk.txt", sep = "\t")
colnames(rmsk_df) <- c("bin", "swScore", "milliDiv", "milliDel", "milliIns", 
                         "genoName", "genoStart", "genoEnd", "genoLeft", "strand", 
                         "repName", "repClass", "repFamily", "repStart",	"repEnd",	
                         "repLeft",	"id")
rmsk_granges <- GRanges(seqnames = rmsk_df$genoName,
                        ranges = IRanges(start = rmsk_df$genoStart,
                                         end = rmsk_df$genoEnd))
rmsk_granges$repFamily <- rmsk_df$repFamily
# Grab a list of all or a subset of consensus peak_files.
# /Shares/rinn_class/data/k562_chip/analysis/00_consensus_peaks/results/
## HINT: list.files("directory", full.names = T)
peak_files <- list.files("/Shares/rinn_class/data/k562_chip/analysis/00_consensus_peaks/results/", full.names = T)
# Extract from these filenames the name of the TF
## Q: what is str_extract(file_name, "^[^_]+(?=_)")
## To understand what this is doing, go to: https://regexr.com/
#tf_names
file_names <- str_extract(peak_files, "[\\w-]+\\.bed")
tf_names <- str_extract(file_names, "^[^_]+(?=_)")
TE_family_names <- unique(rmsk_df$repFamily)
# Produce a list a GRanges one element for each TF peaks.
peak_list <- c()
for(i in 1:length(peak_files)) {
  # Import peaks
  peaks <- rtracklayer::import(peak_files[i])
  # Append this GRanges object to the of the list.
  peak_list <- c(peak_list, peaks)
  # Name the list elements by their TF name.
  names(peak_list)[length(peak_list)] <- tf_names[i]
}
```
## Subset features
Now the goal is to generate a GRanges object (or list of GRanges objects -- i.e., for TE families)
that corresponds to the subset of genomic features that you would like to overlap with your peaks.
```{r subset_features}
# Start from either the gencode gtf GRanges or the RMSK GRanges
# Create a subset of features that are just lncRNA genes or mRNA genes or a repeat family.
# Spend some time looking at the Gencode metatdata.
## Q: For the Gencode file, for a particular gene, how many entries are there?
#nrow(rmsk_df[rmsk_df$repFamily == "Alu",])
## Q: What are the types of the entries for Firre?
#firre_granges <- gtf[gtf$gene_name == "FIRRE"]
#table(firre_granges$type)
# Q: So if we want peaks that intersect witht the gene body, which entries in this
# GRanges do we want?
# Q: What are all the gene_types?
## HINT: gtf$gene_type
# Let's actually make the subset
# Subset the gencode to just lncRNAs|mRNAs|TEs
gypsy_genes <- rmsk_granges[which(rmsk_granges$repFamily == "Gypsy"),]
# Let's export this as a bed file.
```
## Intersect features with peaks
Okay, here's where the real work is done. What we want to end up with is a list of peak sets which
contain just the peaks that overlap with lncRNAs|mRNAs|TEs.
```{r intersect}
gypsy_genes <- rmsk_granges[which(rmsk_granges$repFamily == "Gypsy"),]
alu_genes <- rmsk_granges[which(rmsk_granges$repFamily == "Alu"),]
L1_genes <- rmsk_granges[which(rmsk_granges$repFamily == "L1"),]
snRNA_genes <- rmsk_granges[which(rmsk_granges$repFamily == "snRNA"),]
LTR_genes <- rmsk_granges[which(rmsk_granges$repFamily == "LTR"),]
Satellite_genes <- rmsk_granges[which(rmsk_granges$repFamily == "Satellite"),]
rRNA_genes <- rmsk_granges[which(rmsk_granges$repFamily == "rRNA"),]
Penelope_genes <- rmsk_granges[which(rmsk_granges$repFamily == "Penelope"),]
erv1_genes <- rmsk_granges[which(rmsk_granges$repFamily == "ERV1"),]
# Start with an empty list.
overlaps_list <- c()
# Iterate over the peak list of the full peak sets.
for(i in 1:length(peak_list)) {
  ov <- findOverlaps(peak_list[[i]], gypsy_genes)
  # Fill in the attribute which will get you indices of the peaks that overlap
  overlapping_peaks <- peak_list[[i]][unique(ov@from)]
  # Append to list and make sure it has names.
  overlaps_list <- c(overlaps_list, overlapping_peaks)
  names(overlaps_list)[length(overlaps_list)] <- names(peak_list)[i]
}

```
## Export new peak subset
```{r}
# Export this list of peaks.
## HINT: Naming these something like [tf]_lncRNApeaks.bed would be helpful.
setwd("/scratch/Users/gico3590/CLASS_2020/TE_Overlaps")
#i <- 1
for (i in 1:length(tf_names)) {
  rtracklayer::export(overlaps_list[[i]], paste0(tf_names[i], "_Penelope_peaks.bed"))
}
```
### FIN
If you've gotten here, then you've finished the assigned exercise, but you've also
just arrived at the fun part. Beyond here you can dig into the data and make some plots
to answer questions about this data. We've provided you one figure below
that you can rearrange and use to make the plots you desire.
# Analysis
```{r peak_number}
# This will plot the total number of peaks vs lncrna peaks per tf
# Let's create a data frame of the data we want to plot.
peak_count <- data.frame("tf" = names(peak_list),
  "total_peaks" = sapply(peak_list, length),
  "Alu_peaks" = sapply(overlaps_list, length))
# Here's the actual plot
g <- ggplot(peak_count, aes(x = total_peaks, y = Alu_peaks, label = tf, ymax = 43000, xmax = 62000))
g + geom_point() +
  geom_abline(slope = 1, lty = 2) +
  geom_text_repel(data = peak_count %>% filter(total_peaks > 32000))

# Make sure to save the plot!
```