---
title: "Expression vs. DNA-binding at promoters"
output: html_document
editor_options:
  chunk_output_type: console
---
```{r}
BiocManager::install("regioneR")
```

```{r setup}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = F)

library(tidyverse)
library(GenomicRanges)
library(GenomicFeatures)
library(rtracklayer)
library(ggrepel)
library(Rsubread)
library(regioneR)

source("analysis/util/intersect_functions.R")
source("analysis/util/_setup.R")

```

#### Source for RNA-seq data

# total RNA-seq for K562 dataset: https://www.encodeproject.org/experiments/ENCSR885DVH/

# system("wget https://www.encodeproject.org/files/ENCFF625ZBS/@@download/ENCFF625ZBS.bam",
#       intern = FALSE, ignore.stdout = TRUE, ignore.stderr = TRUE)
# system("wget https://www.encodeproject.org/files/ENCFF630HEX/@@download/ENCFF630HEX.bam",
#        intern = FALSE, ignore.stdout = TRUE, ignore.stderr = TRUE)


```{r read_counts_to_TPM}

counts <- featureCounts(c("ENCFF625ZBS.bam", "ENCFF630HEX.bam"),
                        annot.ext = "../../../Shares/rinn_class/data/genomes/human/gencode/v32/gencode.v32.annotation.gtf",
                        isGTFAnnotationFile = T,
                        isPairedEnd = TRUE,
                        nthreads = 16)

rpk <- counts$counts / (counts$annotation$Length/1000)
expression <- data.frame("rpk" = rpk) %>%
  rownames_to_column("gene_id") %>%
  pivot_longer(2:3, names_to = "sample", values_to = "rpk")

expression_summary <- expression %>%
  group_by(sample) %>%
  summarize(total_rpk = sum(rpk, na.rm = T))
expression_summary$rpk_m <- expression_summary$total_rpk / 1e6
expression <- merge(expression, expression_summary)
expression$tpm <-expression$rpk / expression$rpk_m
tpm <- expression %>% group_by(gene_id) %>%
  summarize(tpm = mean(tpm, na.rm = T))
```


```{r consensus_peaks}

#' @description import_peaks
#' this function will compile .bed files of consensus peaks for multiple DNA-binding proteins into a list
#' 
#' @param consensus_file_path
#'the file path in which the .bed files are located

import_peaks <- function(consensus_file_path = "/Shares/rinn_class/data/CLASS_2020/analysis/01_consensus_peaks/results") {
  peak_files <- list.files(consensus_file_path, full.names = T)
  file_names <- str_extract(peak_files, "[\\w-]+\\.bed")
  tf_name <- str_extract(file_names, "^[^_]+(?=_)")
  
  peak_list <- c()
  for(i in 1:length(peak_files)) {
    peaks <- rtracklayer::import(peak_files[i])
    peak_list <- c(peak_list, peaks)
    names(peak_list)[length(peak_list)] <- tf_name[i]
  }

  return(peak_list)
}

peak_list <- import_peaks("/Shares/rinn_class/data/CLASS_2020/analysis/01_consensus_peaks/results")

```


```{r promoters_from_gencode}

gtf <- rtracklayer::import("/Shares/rinn_class/data/genomes/human/gencode/v32/gencode.v32.annotation.gtf")

#' @description get_promoter_regions
#' this function will create a Granges object of promoter regions from an annotated reference genome
#' 
#' @param gencode_gr
#'  Granges object containing gencode annotated reference genome
#'
#' @param upstream
#'  number of kilobases upstream of a TSS you'd define as a promoter
#'
#' @param downstream
#'  number of kilobases downstream of a TSS you'd define as a promoter

get_promoter_regions <- function(gencode_gr, upstream = 3000, downstream = 3000) {
  genes <- gencode_gr[gencode_gr$type == "gene"]
  proms <- GenomicRanges::promoters(genes, upstream, downstream)
return(proms)
}

gtf_promoters <- get_promoter_regions(gtf)
promoters <- gtf_promoters[which(gtf_promoters$gene_type == "lncRNA" | gtf_promoters$gene_type == "protein_coding")]

```


```{r peaks_per_promoter}

#' @description count_peaks_per_feature
#' this function will create a matrix with counts for all DNA-binding protein peaks in each genomic region
#' 
#' @param features
#'  Granges object containing genomic features of interest
#'
#' @param peak_list
#'  list of consensus peaks

count_peaks_per_feature <- function(features, peak_list) {

  peak_count <- matrix(numeric(), ncol = length(features), nrow = 0)
  
  for(j in 1:length(peak_list)) {
    ov <- countOverlaps(features, peak_list[[j]])
    peak_count <- rbind(peak_count, ov)
    rownames(peak_count)[nrow(peak_count)] <- names(peak_list)[j]
    colnames(peak_count) <- features$gene_id
  }
  return(peak_count)
}

promoter_peak_count <- count_peaks_per_feature(promoters, peak_list)

```


```{r peak_count_data_frame}

promoter_peak_occurence <- matrix(as.numeric(promoter_peak_count > 0),
                           nrow = dim(promoter_peak_count)[1],
                           ncol = dim(promoter_peak_count)[2])
rownames(promoter_peak_occurence) <- rownames(promoter_peak_count)
colnames(promoter_peak_occurence) <- colnames(promoter_peak_count)

peak_occurence_df <- data.frame("gene_id" = colnames(promoter_peak_occurence),
                                "number_of_tfs" = colSums(promoter_peak_occurence),
                                'gene_type'=promoters$gene_type)

peak_occurence_df <- merge(peak_occurence_df,tpm)

```

Figure A: Scatter plot of Number of Transcription factors vs. Expression. The trend lines are based on the number of Transcription Factors (TFs) binding to a gene versus ChiP Peak expression data.  In red are gene that are protein coding.  In black are genes that encode lncRNAs.  
```{r plot_expression_vs_binding_events}

g <- ggplot(peak_occurence_df, aes(y = log10(tpm), x = number_of_tfs, color = gene_type))

g + geom_point(alpha = 0.2) +
  geom_smooth() +
  scale_y_continuous(expand = c(0,0)) +
  scale_x_continuous(expand = c(0,0)) +
  scale_color_manual(values = c("#424242","#a8404c"), name = "Gene type") +
  ggtitle("Expression vs. promoter binding events") +
  xlab(expression('Number of TFs')) +
  ylab(expression(log[10](TPM))) 
  

# ggsave("figures/k562_expression_vs_promoter_binding_all_points.png")
# ggsave("figures/k562_expression_vs_promoter_binding_all_points.pdf")
```

Figure B: Scattter plot of number of transcription factors bound to a promoter vs. promoter Expression. The trend lines are based on the number of Transcription Factors (TFs) binding to a gene versus ChiP Peak expression data.  In red are gene that are protein coding.  In black are genes that encode lncRNAs. The data points not fitted to a trendline have TF binding but not ChiP Peak expression.
```{r plot_expression_v_binding_events_ghosts}

g <- ggplot(peak_occurence_df, aes(y = log10(tpm), x = number_of_tfs, color = gene_type))
g + 
  geom_point(data = peak_occurence_df %>% filter(tpm < 0.001), alpha = 0.2) +
  geom_smooth() + 
  scale_y_continuous(expand = c(0,0)) + 
  scale_x_continuous(expand = c(0,0)) + 
  scale_color_manual(values = c("#424242","#a8404c"), name = "Gene type") + 
  ggtitle("Expression vs. promoter binding events") + 
  xlab(expression('Number of TFs')) +
  ylab(expression(log[10](TPM)))
# ggsave("figures/k562_expression_vs_promoter_binding.png")
# ggsave("figures/k562_expression_vs_promoter_binding.pdf")
```

Figure C: Boxplot of lncRNA and protein coding genes that are expressed and not expressed and their corresponding number of bound TFs. x-values range from 30-100 TFs based on Figure A,B.

Figure D: Histogram of promoter density occurence vs the number_of_tfs. 
```{r biotype_histogram}

peak_occurence_df$expression <- "off"
peak_occurence_df[which(peak_occurence_df$tpm > 0.001), "expression"] <- "expressed"

g <- ggplot(peak_occurence_df, aes(x = number_of_tfs))
g + geom_density(alpha = 0.2, fill = "#424242")  +
  geom_vline(xintercept = 30, lty = 2) +
  geom_vline(xintercept = 100, lty = 2)+
  ggtitle("Promoter Density vs Number of TFs")



g <- ggplot(peak_occurence_df, aes(x = number_of_tfs))
g + geom_histogram(binwidth = 5)  + 
  xlim(30, 100) +
  facet_wrap(expression~gene_type, scales = "free_y") 
  


```

Figure E: Peak occurence vs number of TFs and density of promoters. In blue are genes with low TF binding (# of TF < 25). Red histogram shows genes with high number of TFs (# of TF > 25). x-axis shows transcripts per kilobase million. y-axis show promoter density. 
```{r plot_peak_occurrence}

peak_occurence_df$tf_binding <- "low"
peak_occurence_df[which(peak_occurence_df$number_of_tfs > 25), "tf_binding"] <- "high"

g <- ggplot(peak_occurence_df, aes(x = log10(tpm), color = tf_binding))
g + geom_density(alpha = 0.2) 
```
 
###Chi-square
title: "mRNA vs lncRNA TR binding"
output: html_document
editor_options: 
  chunk_output_type: console
---
```{r setup, include=FALSE}
options(stringsAsFactors = FALSE)
knitr::opts_chunk$set(echo = TRUE)
source("../util/intersect_functions.R")
source("../util/_setup.R")
library(GenomicRanges)
library(tidyverse)
library(regioneR)
```

```{r chi_squared}

# Get ghost promoters
ghosts <- peak_occurence_df %>% filter(tpm < 0.001)
ghosts <- ghosts[which(ghosts$number_of_tfs >= 7),]
ghost_promoters <- gtf_promoters[gtf_promoters$gene_id %in% ghosts$gene_id,]

# Get POLII peaks
poliia <- rtracklayer::import("/Shares/rinn_class/data/CLASS_2020/analysis/01_consensus_peaks/results/POLR2A_consensus_peaks.bed")
poliib <- rtracklayer::import("/Shares/rinn_class/data/CLASS_2020/analysis/01_consensus_peaks/results/POLR2B_consensus_peaks.bed")

# Get ghosts with and without POLII in them
pol_ghost_overlaps <- findOverlaps(ghost_promoters, pollii_peaks)
pol_ghost_proms <- ghost_promoters[unique(pol_ghost_overlaps@from)]
nn <- c(1:length(ghost_promoters))
mm <- subset(nn, !(nn %in% pol_ghost_overlaps@from))
nopol_ghost_proms <- ghost_promoters[mm]

pol_matrix <- count_peaks_per_feature(pol_ghost_proms, peak_list, type = "occurence")
nopol_matrix <- count_peaks_per_feature(nopol_ghost_proms, peak_list, type = "occurence")

pol_num_overlaps <- rowsums(pol_matrix)
nopol_matrix <- rowsums(nopol_matrix)
num_peaks <- sapply(peak_list, length)
total_pol <- length(pol_ghost_proms)
total_nopol <- length(nopol_ghost_proms)
total_ghosts <- length(ghost_promoters)
chisq_res <- data.frame("tf" =character(),
                  "pol_peaks_observed" = numeric(),
                  "chisq_stat" = numeric(),
                  "chisq_pval" = numeric(),
                  "pol_peaks_expected" = numeric())

# from here on out is Michael's code
lncrna_num_overlaps <- rowSums(lncrna_matrix)
mrna_num_overlaps <- rowSums(mrna_matrix)
num_peaks <- sapply(peak_list, length)
total_lncrna <- length(lncrna_promoters)
total_mrna <- length(mrna_promoters)
total_genes <- total_lncrna + total_mrna
chisq_res <- data.frame("tf" =character(),
                  "lncrna_peaks_observed" = numeric(),
                  "chisq_stat" = numeric(),
                  "chisq_pval" = numeric(),
                  "lncrna_peaks_expected" = numeric())
for(i in 1:length(num_peaks)) {
  total_peaks <- num_peaks[[i]]
  df1 <- data.frame("gene_type" = c("lncRNA","lncRNA", "mRNA", "mRNA"),
                    "promoter_bound" = c("bound", "not_bound", "bound", "not_bound"),
                    "count" = c(lncrna_num_overlaps[[i]],
                                total_lncrna - lncrna_num_overlaps[[i]],
                                mrna_num_overlaps[[i]],
                                total_mrna - mrna_num_overlaps[[i]])) %>%
    pivot_wider(names_from = gene_type, values_from = count) %>%
    column_to_rownames("promoter_bound") %>%
    as.matrix()
  
  csres <- chisq.test(df1)
  tdf <- data.frame("tf" = names(num_peaks[i]),
                    "lncrna_peaks_observed" = lncrna_num_overlaps[[i]],
                    "chisq_stat" = csres$statistic,
                    "chisq_pval" = csres$p.value,
                    "lncrna_peaks_expected" = csres$expected[1,1])
  chisq_res <- bind_rows(chisq_res, tdf)
}
chisq_res$padj <- p.adjust(chisq_res$chisq_pval, method = "BH")
chisq_res$diff <- chisq_res$lncrna_peaks_observed - chisq_res$lncrna_peaks_expected
write_csv(chisq_res, "results/lncrna_vs_mrna_chisq_results.csv")
# contingency table
# lncRNA yes lncRNA no
# tf bound
# tf not bound
```

