---
title: "Expression vs. DNA-binding at promoters"
output: html_document
editor_options:
  chunk_output_type: console
---

```{r setup}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = F)

library(tidyverse)
library(GenomicRanges)
library(GenomicFeatures)
library(rtracklayer)
library(ggrepel)
library(Rsubread)
library(regioneR)

source("../util/intersect_functions.R")
source("../util/_setup.R")

```


# Purpose: Compare expression of genes to binding in promoter regions

This script extracts expression counts from K562 RNA-seq data and identifies DNA-binding events at promoters from the K562 ChIP-seq consensus peak dataset. We have included further analysis on promoters that contain â‰¥7 DNA-binding events but are not expressed.


### Source for RNA-seq data

total RNA-seq for K562 dataset: https://www.encodeproject.org/experiments/ENCSR885DVH/

system("wget https://www.encodeproject.org/files/ENCFF625ZBS/@@download/ENCFF625ZBS.bam",
      intern = FALSE, ignore.stdout = TRUE, ignore.stderr = TRUE)
system("wget https://www.encodeproject.org/files/ENCFF630HEX/@@download/ENCFF630HEX.bam",
       intern = FALSE, ignore.stdout = TRUE, ignore.stderr = TRUE)


### Analysis of RNA-seq data

```{r read_counts_to_TPM}

counts <- featureCounts(c("ENCFF625ZBS.bam", "ENCFF630HEX.bam"),
                        annot.ext = "../../../../../Shares/rinn_class/data/genomes/human/gencode/v32/gencode.v32.annotation.gtf",
                        isGTFAnnotationFile = T,
                        isPairedEnd = TRUE,
                        nthreads = 16)

rpk <- counts$counts / (counts$annotation$Length/1000)
expression <- data.frame("rpk" = rpk) %>%
  rownames_to_column("gene_id") %>%
  pivot_longer(2:3, names_to = "sample", values_to = "rpk")

expression_summary <- expression %>%
  group_by(sample) %>%
  summarize(total_rpk = sum(rpk, na.rm = T))
expression_summary$rpk_m <- expression_summary$total_rpk / 1e6
expression <- merge(expression, expression_summary)
expression$tpm <-expression$rpk / expression$rpk_m
tpm <- expression %>% group_by(gene_id) %>%
  summarize(tpm = mean(tpm, na.rm = T))
```


### Import consensus peak set

For details on how the consensus peak set was determined, see "CLASS_2020/analysis/01_consensus_peaks/consensus_peak_set.Rmd"

```{r consensus_peaks}

#' @description import_peaks
#' this function will compile .bed files of consensus peaks for multiple DNA-binding proteins into a list
#' 
#' @param consensus_file_path
#'the file path in which the .bed files are located

import_peaks <- function(consensus_file_path = "/Shares/rinn_class/data/CLASS_2020/analysis/01_consensus_peaks/results") {
  peak_files <- list.files(consensus_file_path, full.names = T)
  file_names <- str_extract(peak_files, "[\\w-]+\\.bed")
  tf_name <- str_extract(file_names, "^[^_]+(?=_)")
  
  peak_list <- c()
  for(i in 1:length(peak_files)) {
    peaks <- rtracklayer::import(peak_files[i])
    peak_list <- c(peak_list, peaks)
    names(peak_list)[length(peak_list)] <- tf_name[i]
  }

  return(peak_list)
}

peak_list <- import_peaks()

```

### Import promoter regions from Gencode for lncRNA- and mRNA-coding genes

```{r promoters_from_gencode}

gtf <- rtracklayer::import("/Shares/rinn_class/data/genomes/human/gencode/v32/gencode.v32.annotation.gtf")

#' @description get_promoter_regions
#' this function will create a Granges object of promoter regions from an annotated reference genome
#' 
#' @param gencode_gr
#'  Granges object containing gencode annotated reference genome
#'
#' @param upstream
#'  number of kilobases upstream of a TSS you'd define as a promoter
#'
#' @param downstream
#'  number of kilobases downstream of a TSS you'd define as a promoter

get_promoter_regions <- function(gencode_gr, upstream = 3000, downstream = 3000) {
  genes <- gencode_gr[gencode_gr$type == "gene"]
  proms <- GenomicRanges::promoters(genes, upstream, downstream)
return(proms)
}

gtf_promoters <- get_promoter_regions(gtf)
promoters <- gtf_promoters[which(gtf_promoters$gene_type == "lncRNA" | gtf_promoters$gene_type == "protein_coding")]

```


### Count consensus peaks in each promoter region

```{r peaks_per_promoter}

#' @description count_peaks_per_feature
#' this function will create a matrix with counts for all DNA-binding protein peaks in each genomic region
#' 
#' @param features
#'  Granges object containing genomic features of interest
#'
#' @param peak_list
#'  list of consensus peaks

count_peaks_per_feature <- function(features, peak_list) {

  peak_count <- matrix(numeric(), ncol = length(features), nrow = 0)
  
  for(j in 1:length(peak_list)) {
    ov <- countOverlaps(features, peak_list[[j]])
    peak_count <- rbind(peak_count, ov)
    rownames(peak_count)[nrow(peak_count)] <- names(peak_list)[j]
    colnames(peak_count) <- features$gene_id
  }
  return(peak_count)
}

promoter_peak_count <- count_peaks_per_feature(promoters, peak_list)

```


```{r peak_count_data_frame}

promoter_peak_occurrence <- matrix(as.numeric(promoter_peak_count > 0),
                           nrow = dim(promoter_peak_count)[1],
                           ncol = dim(promoter_peak_count)[2])
rownames(promoter_peak_occurrence) <- rownames(promoter_peak_count)
colnames(promoter_peak_occurrence) <- colnames(promoter_peak_count)

peak_occurrence_df <- data.frame("gene_id" = colnames(promoter_peak_occurrence),
                                "number_of_tfs" = colSums(promoter_peak_occurrence),
                                'gene_type'=promoters$gene_type)

peak_occurrence_df <- merge(peak_occurrence_df,tpm)

write_csv(peak_occurrence_df, "results/k562_expression_binding.csv")

```


### Chi-square statistics for reservoir promoters with POLII binding vs. without POLII binding

```{r chi_squared}

# Get ghost promoters
ghosts <- peak_occurrence_df %>% filter(tpm < 0.001)
ghosts <- ghosts[which(ghosts$number_of_tfs > 7),]
ghost_promoters <- gtf_promoters[gtf_promoters$gene_id %in% ghosts$gene_id,]

# Get peaks
poliia <- rtracklayer::import("/Shares/rinn_class/data/CLASS_2020/analysis/01_consensus_peaks/results/POLR2A_consensus_peaks.bed")
poliib <- rtracklayer::import("/Shares/rinn_class/data/CLASS_2020/analysis/01_consensus_peaks/results/POLR2B_consensus_peaks.bed")
polii_peaks <- c(poliia, poliib)

# Get ghosts with and without POLII in them
pol_ghost_overlaps <- findOverlaps(ghost_promoters, polii_peaks)
pol_ghost_proms <- ghost_promoters[unique(pol_ghost_overlaps@from)]
nn <- c(1:length(ghost_promoters))
mm <- subset(nn, !(nn %in% pol_ghost_overlaps@from))
nopol_ghost_proms <- ghost_promoters[mm]

# This is a rehash of the count_peaks_per_feature function so it will work with the "occurrence" option.

count_peaks_per_feature <- function(features, peak_list, type) {
  
  if(!(type %in% c("counts", "occurrence"))) {
    stop("Type must be either occurrence or counts.")
  }
  
  peak_count <- matrix(numeric(), ncol = length(features), nrow = 0)
  
  for(j in 1:length(peak_list)) {
    ov <- countOverlaps(features, peak_list[[j]])
    peak_count <- rbind(peak_count, ov)
    rownames(peak_count)[nrow(peak_count)] <- names(peak_list)[j]
    colnames(peak_count) <- features$gene_id
  }
  
  peak_matrix <- peak_count
  
  if(type == "occurrence") {
    peak_occurrence <- matrix(as.numeric(peak_count > 0), 
                             nrow = dim(peak_count)[1],
                             ncol = dim(peak_count)[2])
    rownames(peak_occurrence) <- rownames(peak_count)
    colnames(peak_occurrence) <- colnames(peak_count)
    peak_matrix <- peak_occurrence
  }
  
  return(peak_matrix)
  
}
pol_matrix <- count_peaks_per_feature(pol_ghost_proms, peak_list, type = "occurrence")
nopol_matrix <- count_peaks_per_feature(nopol_ghost_proms, peak_list, type = "occurrence")

pol_num_overlaps <- rowSums(pol_matrix)
nopol_num_overlaps <- rowSums(nopol_matrix)
num_peaks <- sapply(peak_list, length)
total_pol <- length(pol_ghost_proms)
total_nopol <- length(nopol_ghost_proms)
total_ghosts <- length(ghost_promoters)

chisq_res <- data.frame("tf" =character(),
                  "polii_peaks_observed" = numeric(),
                  "chisq_stat" = numeric(),
                  "chisq_pval" = numeric(),
                  "polii_peaks_expected" = numeric())

for(i in 1:length(num_peaks)) {
  total_peaks <- num_peaks[[i]]
  df1 <- data.frame("gene_type" = c("pol","pol", "nopol", "nopol"),
                    "promoter_bound" = c("bound", "not_bound", "bound", "not_bound"),
                    "count" = c(pol_num_overlaps[[i]],
                                total_pol - pol_num_overlaps[[i]],
                                nopol_num_overlaps[[i]],
                                total_nopol - nopol_num_overlaps[[i]])) %>%
    pivot_wider(names_from = gene_type, values_from = count) %>%
    column_to_rownames("promoter_bound") %>%
    as.matrix()
  csres <- chisq.test(df1)
  tdf <- data.frame("tf" = names(num_peaks[i]),
                    "polii_peaks_observed" = pol_num_overlaps[[i]],
                    "chisq_stat" = csres$statistic,
                    "chisq_pval" = csres$p.value,
                    "polii_peaks_expected" = csres$expected[1,1])
  chisq_res <- bind_rows(chisq_res, tdf)
}
chisq_res$padj <- p.adjust(chisq_res$chisq_pval, method = "BH")
chisq_res$diff <- chisq_res$polii_peaks_observed - chisq_res$polii_peaks_expected

write_csv(chisq_res, "results/k562_polII_reservoirs_chisquare.csv")

```


### Make plots for figures

Figure A: Scatter plot of expression vs. number of DNA-binding factors bound for lncRNA and protein-coding promoters 
```{r plot_expression_vs_binding_events}

g <- ggplot(peak_occurrence_df, aes(y = log10(tpm), x = number_of_tfs, color = gene_type))

g + geom_point(alpha = 0.2) +
  geom_smooth() +
  scale_y_continuous(expand = c(0,0)) +
  scale_x_continuous(expand = c(0,0)) +
  scale_color_manual(values = c("#424242","#a8404c"), name = "Gene type") +
  ggtitle("Expression vs. promoter binding events") +
  xlab(expression('Number of TFs')) +
  ylab(expression(log[10](TPM))) 

ggsave("figures/k562_expression_vs_promoter_binding_all_points.png")
ggsave("figures/k562_expression_vs_promoter_binding_all_points.pdf")
```


Figure B: Scatter plot of expression vs. number of DNA-binding factors bound for lncRNA and protein-coding promoters. Promoters with >7 factors bound but no expression are shown as dots.
```{r plot_expression_v_binding_events_ghosts}

g <- ggplot(peak_occurrence_df, aes(y = log10(tpm), x = number_of_tfs, color = gene_type))
g + 
  geom_point(data = peak_occurrence_df %>% filter(tpm < 0.001), alpha = 0.2) +
  geom_smooth() + 
  scale_y_continuous(expand = c(0,0)) + 
  scale_x_continuous(expand = c(0,0)) + 
  scale_color_manual(values = c("#424242","#a8404c"), name = "Gene type") + 
  ggtitle("Expression vs. promoter binding events") + 
  xlab(expression('Number of TFs')) +
  ylab(expression(log[10](TPM)))

ggsave("figures/k562_expression_vs_promoter_binding.png")
ggsave("figures/k562_expression_vs_promoter_binding.pdf")
```


Figure C: Histogram of promoters by number of DNA-binding factors bound
```{r binding_histogram}

peak_occurrence_df$expression <- "off"
peak_occurrence_df[which(peak_occurrence_df$tpm > 0.001), "expression"] <- "expressed"

g <- ggplot(peak_occurrence_df, aes(x = number_of_tfs))
g + geom_density(alpha = 0.2, fill = "#424242")  +
  geom_vline(xintercept = 30, lty = 2) +
  geom_vline(xintercept = 100, lty = 2)+
  ggtitle("Promoter Density vs Number of TFs")

ggsave("figures/k562_binding_histogram.png")
ggsave("figures/k562_binding_histogram.pdf")

```


Figure D: Histogram of promoters by number of DNA-binding factors bound for different biotypes and expression levels
```{r expression_histogram}

g <- ggplot(peak_occurrence_df, aes(x = number_of_tfs))
g + geom_histogram(binwidth = 5)  + 
  xlim(30, 100) +
  facet_wrap(expression~gene_type, scales = "free_y")
  
ggsave("figures/k562_binding_histogram_expression.png")
ggsave("figures/k562_binding_histogram_expression.pdf")

```


Figure E: Histogram of promoters by gene expression for different DNA-binding factor densities. In red are prommoters with fewer than 25 factors bound). In black are promomters with 25 or more factors bound.
```{r plot_peak_occurrence}

peak_occurrence_df$tf_binding <- "low"
peak_occurrence_df[which(peak_occurrence_df$number_of_tfs > 25), "tf_binding"] <- "high"

g <- ggplot(peak_occurrence_df, aes(x = log10(tpm), color = tf_binding))
g + geom_density(alpha = 0.2) +
  scale_color_manual(values = c("#424242","#a8404c"), name = "tf_binding")

ggsave("figures/k562_expression_histogram.png")
ggsave("figures/k562_expression_histogram.pdf")

```


Figure F: Chi-square values vs. difference in expected and observed binding for DNA-binding factors at promoters with or without POLII bound
```{r plot_chi_square}

g <- ggplot(chisq_res, aes(x = diff, y = -log10(padj)))
g + geom_point(colour = "#424242") +
    geom_text_repel(
    data = subset(chisq_res, diff > 100 | diff < -10),
    aes(label = tf),
    size = 3,
    color ="#a8404c",
    box.padding = unit(0.35, "lines"),
    point.padding = unit(0.3, "lines"))

ggsave("figures/k562_polII_reservoirs_chisquare.png")
ggsave("figures/k562_polII_reservoirs_chisquare.pdf")

```

