---
title: "K562 RNA-seq expression"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE)
library(Rsubread)
```

```{r}
# system("wget https://www.encodeproject.org/files/ENCFF625ZBS/@@download/ENCFF625ZBS.bam",
#        intern = FALSE, ignore.stdout = TRUE, ignore.stderr = TRUE)
# system("wget https://www.encodeproject.org/files/ENCFF630HEX/@@download/ENCFF630HEX.bam", 
#        intern = FALSE, ignore.stdout = TRUE, ignore.stderr = TRUE)
counts <- featureCounts(c("ENCFF625ZBS.bam", "ENCFF630HEX.bam"),
                        annot.ext = "/Shares/rinn_class/data/genomes/human/gencode/v32/gencode.v32.annotation.gtf",
                        isGTFAnnotationFile = T,
                        isPairedEnd = TRUE,
                        nthreads = 16)
```

```{r}
rpk <- counts$counts / (counts$annotation$Length/1000)
expression <- data.frame("rpk" = rpk) %>%
  rownames_to_column("gene_id") %>%
  pivot_longer(2:3, names_to = "sample", values_to = "rpk")
expression_summary <- expression %>% 
  group_by(sample) %>%
  summarize(total_rpk = sum(rpk, na.rm = T))
expression_summary$rpk_m <- expression_summary$total_rpk / 1e6

expression <- merge(expression, expression_summary)

expression$tpm <-expression$rpk / expression$rpk_m

tpm <- expression %>% group_by(gene_id) %>%
  summarize(tpm = mean(tpm, na.rm = T))

gencode_gr <- rtracklayer::import("/Shares/rinn_class/data/genomes/human/gencode/v32/gencode.v32.annotation.gtf")
genes <- gencode_gr[gencode_gr$type == "gene"] %>% as.data.frame() %>%
  dplyr::select(gene_name, gene_id)
tpm <- merge(tpm, genes)
tpm <- tpm %>% dplyr::select(gene_id, gene_name, tpm)
write_csv(tpm, "results/k562_tpm.csv")
```


