---
title: "Density plot for DNA-binding versus expression at promoters"
output: html_document
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
options(stringsAsFactors = F)
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
library(GenomicFeatures)
library(rtracklayer)
library(ggrepel)
library(Rsubread)
```

#### Expression data obtained here:
total RNA-seq for K562 dataset: https://www.encodeproject.org/experiments/ENCSR885DVH/

system("wget https://www.encodeproject.org/files/ENCFF625ZBS/@@download/ENCFF625ZBS.bam",
        intern = FALSE, ignore.stdout = TRUE, ignore.stderr = TRUE)
        
system("wget https://www.encodeproject.org/files/ENCFF630HEX/@@download/ENCFF630HEX.bam",
        intern = FALSE, ignore.stdout = TRUE, ignore.stderr = TRUE)


Obtain expression data to correlate with DNA-binding.
```{r converting read counts to TPM}

# First we will take the bam file and find the number of reads that align over mRNA and lncRNA gene-bodies
counts <- featureCounts(c("/scratch/Users/gico3590/bchm5631/expression_data/ENCFF625ZBS.bam", "/scratch/Users/gico3590/bchm5631/expression_data/ENCFF630HEX.bam"),
                        annot.ext = "/Shares/rinn_class/data/genomes/human/gencode/v32/gencode.v32.annotation.gtf",
                        isGTFAnnotationFile = T,
                        isPairedEnd = TRUE,
                        nthreads = 16)

# Here we will find the number of reads per kilobases of a given gene body (rpkm).
rpk <- counts$counts / (counts$annotation$Length/1000)
expression <- data.frame("rpk" = rpk) %>%
  rownames_to_column("gene_id") %>%
  pivot_longer(2:3, names_to = "sample", values_to = "rpk")

# Here we will convert rpkm to transcripts per million of reads (tpm)
expression_summary <- expression %>%
  group_by(sample) %>%
  summarize(total_rpk = sum(rpk, na.rm = T))
expression_summary$rpk_m <- expression_summary$total_rpk / 1e6
expression <- merge(expression, expression_summary)
expression$tpm <-expression$rpk / expression$rpk_m
tpm <- expression %>% group_by(gene_id) %>%
  summarize(tpm = mean(tpm, na.rm = T))
```


Obtain promoter regions for lncRNA and protein-coding genes.
```{r}
#Obtain necessary functions and plot themes
source("../util/intersect_functions.R")
source("../util/_setup.R")

# Read in consensus peaks as a list.
# import_peaks is a function
peak_list <- import_peaks()

# Read in gencode genome annotation.
gtf <- rtracklayer::import("/Shares/rinn_class/data/genomes/human/gencode/v32/gencode.v32.annotation.gtf")

# Get the promoter regions of each gene.
# get_promoter_regions is a function
proms_lncRNA <- get_promoter_regions(gtf, biotype = c("lncRNA"))
proms_protein <- get_promoter_regions(gtf, biotype = c("protein_coding"))
proms <- c(proms_lncRNA, proms_protein)

# Count peaks over promoters
# count_peaks_per_feature is a function
promoter_peak_count <- count_peaks_per_feature(proms, peak_list)
```


```{r}
# Make this binary -- a co-occurence matrix.
promoter_peak_occurence <- matrix(as.numeric(promoter_peak_count > 0),
                           nrow = dim(promoter_peak_count)[1],
                           ncol = dim(promoter_peak_count)[2])
rownames(promoter_peak_occurence) <- rownames(promoter_peak_count)
colnames(promoter_peak_occurence) <- colnames(promoter_peak_count)

#Pull gene IDs and transcript types out of gencode gtf
gene_id <- gtf$gene_id
transcript_type <- gtf$transcript_type

#Make a data frame with gene IDs and transcript types.
gene_type_df <- data.frame('gene_id' = gene_id, 'transcript_type' = transcript_type)
gene_type_df <- gene_type_df[which(gene_type_df$transcript_type == c("lncRNA", "protein_coding")),]

# Make a data frame of binding events per promoter.
peak_occurence_df <- data.frame("gene_id" = colnames(promoter_peak_occurence),
                                "number_of_tfs" = colSums(promoter_peak_occurence))

# Merge in the expression data into the data frame you just created.
peak_occurence_df <- merge(peak_occurence_df, gene_type_df)
peak_occurence_df <- merge(peak_occurence_df, tpm)
```


```{r}
# Filter your data frame to promoters that have more than x binding events yet are still not expressed.
# Set x to 1
x <- 1

# Filter peak occurence data frame for promoters that are not expressed (tpm < 0.001)
low_expression <- filter(peak_occurence_df, tpm < 0.001, peak_occurence_df$number_of_tfs > x)

# Filter peak occurence data frame for promoters that are expressed (tpm > 0.001)
expressed <- filter(peak_occurence_df, tpm > 0.001, peak_occurence_df$number_of_tfs > x)

#Filter out only lncRNA promoters which are expressed
lncRNA_expressed <- filter(expressed, expressed$transcript_type == "lncRNA")

#Filter out only lncRNA promoters which are not expressed
lncRNA_not_expressed <- filter(low_expression, low_expression$transcript_type == "lncRNA")

#Filter out only protein-coding promoters which are expressed
protein_coding_expressed <- filter(expressed, expressed$transcript_type == "protein_coding")

#Filter out only protein-coding promoters which are not expressed
protein_coding_not_expressed <- filter(low_expression, low_expression$transcript_type == "protein_coding")
```


```{r}
#Make a density plot for TF binding at lncRNA promoters which are expressed
g <- ggplot(lncRNA_expressed, aes(x = number_of_tfs))
g + geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
  theme_paperwhite() +
  xlab("Number of TFs bound") +
    theme(legend.title = element_text(size = 5), legend.text = element_text(size = 5), legend.key.size = unit(0.5, "cm"), plot.title = element_text(size = 10), axis.title = element_text(size = 7), axis.text.x = element_text(size = 5), axis.text.y = element_text(size = 5)) +
  ggtitle("Density of TFs at expressed lncRNA promoters")
```

Figure 1. Density plot of transcription factors bound to promoters of lncRNA genes that are expressed.

```{r}
ggsave("density_tfs_lncRNAs_expressed.pdf")
ggsave("density_tfs_lncRNAs_expressed.png")

```


```{r}
#Make a density plot for TF binding at lncRNA promoters which are not expressed
g <- ggplot(lncRNA_not_expressed, aes(x = number_of_tfs))
g + geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
  theme_paperwhite() +
  xlab("Number of TFs bound") +
    theme(legend.title = element_text(size = 5), legend.text = element_text(size = 5), legend.key.size = unit(0.5, "cm"), plot.title = element_text(size = 10), axis.title = element_text(size = 7), axis.text.x = element_text(size = 5), axis.text.y = element_text(size = 5)) +
  ggtitle("Density of TFs at not expressed lncRNA promoters")
```

Figure 2. Density plot of transcription factors bound to promoters of lncRNA genes that are not expressed.

```{r}
ggsave("density_tfs_lncRNAs_not_expressed.pdf")
ggsave("density_tfs_lncRNAs_not_expressed.png")
```

```{r}
#Make a density plot for TF binding at protein-coding promoters which are expressed
g <- ggplot(protein_coding_expressed, aes(x = number_of_tfs))
g + geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
  theme_paperwhite() +
  xlab("Number of TFs bound") +
    theme(legend.title = element_text(size = 5), legend.text = element_text(size = 5), legend.key.size = unit(0.5, "cm"), plot.title = element_text(size = 10), axis.title = element_text(size = 7), axis.text.x = element_text(size = 5), axis.text.y = element_text(size = 5)) +
  ggtitle("Density of TFs at expressed protein-coding promoters")
```

Figure 3. Density plot of transcription factors bound to promoters of protein-coding genes that are expressed.

```{r}
ggsave("density_tfs_protein_coding_expressed.pdf")
ggsave("density_tfs_protein_coding_expressed.png")

```

```{r}
#Make a density plot for TF binding at protein-coding promoters which are not expressed
g <- ggplot(protein_coding_not_expressed, aes(x = number_of_tfs))
g + geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
  theme_paperwhite() +
  xlab("Number of TFs bound") +
    theme(legend.title = element_text(size = 5), legend.text = element_text(size = 5), legend.key.size = unit(0.5, "cm"), plot.title = element_text(size = 10), axis.title = element_text(size = 7), axis.text.x = element_text(size = 5), axis.text.y = element_text(size = 5)) +
  ggtitle("Density of TFs at not expressed protein-coding promoters")
```

Figure 4. Density plot of transcription factors bound to promoters of protein-coding genes that are not expressed.

```{r}
ggsave("density_tfs_protein_coding_not_expressed.pdf")
ggsave("density_tfs_protein_coding_not_expressed.png")

```