---
title: "DNA-binding versus expression at promoters"
output: html_document
editor_options:
  chunk_output_type: console
---


```{r setup, include=FALSE}
options(stringsAsFactors = F)
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
library(GenomicFeatures)
library(rtracklayer)
library(ggbio)
library(ggrepel)
library(Rsubread)
```

# How do DNA binding events in turn regulate expression at promoters?
The goal for today is to investigate how DNA binding co-occurences and events correlates with gene-expression levels (RNA-seq)
We won't have time to start with raw RNAsequencing files, but will start pretty close with "read counts".
This is the number of reads identified in a gene-transcript (lncRNA or mRNA) starting from a bam file of alingned reads -- the using Rsubread (a kin to SAM/BAM tools) to count the reads over the transcript.
In the end this first set of code will give us a value for the number of reads and we will eventually convert that to Transcripts Per Million Reads (TPM)

### Goals of this exercise:
1) Create an object with promoters and their correlated expression.
2) What is the distribution of binding events (hint hist())?
3) Design a data frame of promoters that have more than 7 binding events yet are still not expressed.
4) Develop an object of promoters that are not expressed (TPM = 0).


#### This is where the data came from

total RNA-seq for K562 dataset: https://www.encodeproject.org/experiments/ENCSR885DVH/

 system("wget https://www.encodeproject.org/files/ENCFF625ZBS/@@download/ENCFF625ZBS.bam",
       intern = FALSE, ignore.stdout = TRUE, ignore.stderr = TRUE)
system("wget https://www.encodeproject.org/files/ENCFF630HEX/@@download/ENCFF630HEX.bam",
intern = FALSE, ignore.stdout = TRUE, ignore.stderr = TRUE)
```{r}

### This is where the data came from total RNA-seq for K562 dataset: https://www.encodeproject.org/experiments/ENCSR885DVH/
### Make sure your in the right analysis working directory before you run this !

system("wget https://www.encodeproject.org/files/ENCFF625ZBS/@@download/ENCFF625ZBS.bam",
       intern = FALSE, ignore.stdout = TRUE, ignore.stderr = TRUE)
system("wget https://www.encodeproject.org/files/ENCFF630HEX/@@download/ENCFF630HEX.bam",
intern = FALSE, ignore.stdout = TRUE, ignore.stderr = TRUE)

```


```{r converting read counts to TPM}

# First we will take the bam file and find the number of reads that align over mRAN and lncRNA gene-bodies (similar to what we used in MACs)
# Don't worry about the syntax of the following code as it pertains to RNAseq, but maybe worth a google :)

counts <- featureCounts(c("ENCFF625ZBS.bam", "ENCFF630HEX.bam"),
                        annot.ext = "/Shares/rinn_class/data/genomes/human/gencode/v32/gencode.v32.annotation.gtf",
                        isGTFAnnotationFile = T,
                        isPairedEnd = TRUE,
                        nthreads = 16)

# First we will find the number of reads per kilobases of a given gene body (rpkm).
# Q: Why would we want to normalize to the size of the genebody?
rpk <- counts$counts / (counts$annotation$Length/1000)
expression <- data.frame("rpk" = rpk) %>%
  rownames_to_column("gene_id") %>%
  pivot_longer(2:3, names_to = "sample", values_to = "rpk")

# Here we will convert rpkm to transcripts per million of reads (tpm)
# Q: why would we need to furhter normalize this way?
expression_summary <- expression %>%
  group_by(sample) %>%
  summarize(total_rpk = sum(rpk, na.rm = T))
expression_summary$rpk_m <- expression_summary$total_rpk / 1e6
expression <- merge(expression, expression_summary)
expression$tpm <-expression$rpk / expression$rpk_m
tpm <- expression %>% group_by(gene_id) %>%
  summarize(tpm = mean(tpm, na.rm = T))
```


```{r}
#######
## This is not a part of the exercise, but if you want to try it out, it's here.
#######
# let's look at the sample concordance
# library(ggpubr)
# g <- ggplot(tpm , aes(x = log10(ENCFF625ZBS), y = log10(ENCFF630HEX)))
# g + geom_point() + stat_cor()
```

## Class excercise
1) develop an object of promoters that are not expressed (TPM = 0)
2) what is the distribution of binding events (hint hist())
3) design a data frame of promoters that have more than 7 binding events yet are still not expressed



```{r}
####
# HINT: You should use the functions available in util/intersect_functions.R (import_peaks, etc.)
####

# 1. Read in consensus peaks as a list.

# 2. Read in gencode.

# 3. Get the promoter regions of each gene.
# Q: Which gene types should you grab here? Can you input a list of multiple gene types?

# 4. Count peaks over promoters
# HINT: You need to name this variable "promoter_peak_count" for it to work with the next
# bit of code.

```

```{r}
# Make this binary -- a co-occurence matrix.
promoter_peak_occurence <- matrix(as.numeric(promoter_peak_count > 0),
                           nrow = dim(promoter_peak_count)[1],
                           ncol = dim(promoter_peak_count)[2])
rownames(promoter_peak_occurence) <- rownames(promoter_peak_count)
colnames(promoter_peak_occurence) <- colnames(promoter_peak_count)


# Make a data frame of binding events per promoter.
peak_occurence_df <- data.frame("gene_id" = colnames(promoter_peak_occurence),
                                "number_of_tfs" = colSums(promoter_peak_occurence))

# Merge in the expression data into the data frame you just created.
# HINT: ?merge

```


```{r}
# Use what you just made to plot the results.
g <- ggplot(peak_occurence_df, aes(y = log10(tpm), x = number_of_tfs, color = gene_type))
g + geom_point(alpha = 0.2) +
  geom_smooth() +
  scale_y_continuous(expand = c(0,0)) +
  scale_x_continuous(expand = c(0,0)) +
  scale_color_manual(values = c("#424242","#a8404c"), name = "Gene type") +
  ggtitle("Promoter binding events vs. expression") +
  xlab(expression('Number of TFs')) +
  ylab(expression(log[10](TPM)))
ggsave("figures/k562_promoter_binding_vs_expression.png")
ggsave("figures/k562_promoter_binding_vs_expression.pdf")
```



```{r}
# 2) What is the distribution of binding events (hint: hist())?
# At what point would you make a cutoff of "promoters with a lot of TFs vs
# those with few"?



```
```{r}
# With this data.frame, plot the distribution of expression across those two groups.

# For example...
# X should be your cutoff from above.
peak_occurence_df$tf_binding <- "low"
peak_occurence_df[which(peak_occurence_df$number_of_tfs > x), "tf_binding"] <- "high"

g <- ggplot(peak_occurence_df, aes(x = log10(tpm), color = tf_binding))
g + geom_density(alpha = 0.2)
```

```{r}
# 3) Filter your data frame to promoters that have more than X binding events
#    yet are still not expressed.
# HINT: ?filter(tpm < 0.001)

# Do these promoters still have TFs bound? How many?
# Create a plot to illustrate this.


```


```{r}
# Please make any other plots to satisfy your curiosity!
# Let us know what you find!



```


```{r}
# Here's an example of a binned approach.
## Are these reasonable break points? How would consider changing them?
quantile(peak_occurence_df$tpm)
peak_occurence_df$expression <- cut(peak_occurence_df$tpm, breaks = c(-1, 0.001, 0.137, 3, 5000),
                                    labels = c("off", "low", "med", "high"))
table(peak_occurence_df$expression)

g <- ggplot(peak_occurence_df, aes(x = expression, y = number_of_tfs, color = gene_type))
g + geom_boxplot() + stat_compare_means() +
  scale_color_manual(values = c("#a8404c", "#424242"))
```

