---
title: "Clustering by Features"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
options(stringsAsFactors = FALSE)
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
source("../util/intersect_functions.R")
source("../util/_setup.R")
```


# The first step is to get a matrix of features by TFs.

```{r}
# We are going to start out with a function that we use all the time to import files called "import_peaks"
#? What is it doing and what is the output format? See util folder on github

peak_list <- import_peaks("/Shares/rinn_class/data/k562_chip/analysis/00_consensus_peaks/results/")

# Making the features (gencode) GRanges that has lots of features annotated
gencode_gr <- rtracklayer::import("/Shares/rinn_class/data/genomes/human/gencode/v32/gencode.v32.annotation.gtf")

# Here we are using the function "get_promoter_regions" to grab both the lncRNA and mRNA genes
#? What is this funciton doing (see util folder for function file)
promoters <- get_promoter_regions(gencode_gr, biotype = c("?", "?"))

# here we are using the function "count_peaks_per_feature
#? What is it doing (see util folder for function file)
promoter_peak_count <- count_peaks_per_feature(?, ?)
# What did you end up with ? Go over the resulting gencode_gr and peak_list
# What is the distribution of DNA binding events across promoters that had at least 1 event?
# Todo: Make two plots one for mRNA and one for lncRNA


```

```{r filtering}
# Now we want to remove any columns (e.g., promoters) where a TF 
promoter_peak_count <- promoter_peak_count[,which(colSums(promoter_peak_count) > 0)]

# Remove TFs with no promoter peaks
no_promoter_binding <- rownames(promoter_peak_count)[which(rowSums(promoter_peak_count) == 0)]
promoter_peak_count <- promoter_peak_count[which(rowSums(promoter_peak_count) > 0),]

# Make this binary -- make it a co-occurence matrix.
promoter_peak_occurence <- matrix(as.numeric(promoter_peak_count > 0), 
                           nrow = dim(promoter_peak_count)[1],
                           ncol = dim(promoter_peak_count)[2])
rownames(promoter_peak_occurence) <- rownames(promoter_peak_count)
colnames(promoter_peak_occurence) <- colnames(promoter_peak_count)
```

# Now that we have a binary matrix and counts of number of features that intersect with DNA binding events -- let's cluster to see if the grouping makes sense
# We will spare you the time but on the class wiki there is a manuscript on clustering. Since we are using a binary matrix we will calculate the distance betweeen each sample using "dist(binary" 
# Read through the manual page to see other paramaters of this funciton ?dist
```{r}
#TODO clean up 
# Calculate a distance measure between the rows.

bin_hier <- hclust(dist(promoter_peak_occurence, method = "binary"))
plot(bin_hier)
# load library for ggdendro for the best trees we could find :) 
library(ggdendro)

hclust_euc <- hclust(dist(promoter_peak_occurence, method = "euclidean"))
ggdendro::ggdendrogram(hclust_euc, rotate = TRUE,  size = 2)
ggsave("hclust_binary_dist.pdf", height = 26, width = 6)


ggdendro::ggdendrogram(bin_hier, rotate = TRUE,  size = 2)
ggsave("hclust_binary_dist.pdf", height = 26, width = 6)


```

# Class excercise cluster by your group










