---
title: "Peak-Promoter Co-occurence matrix"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
options(stringsAsFactors = FALSE)
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
```

#### Goal: to build functions to import_peaks, intersect them with your favorite genomic feature and turn that into a co-occurance matrix for numerous downstream analyses.
Wednesday present a cool lncRNA and mRNA promoter and any other intersting results you can find in the feature count matrix. 
In otherwords we are bringing it all back together with functions we will use often. 
1) Peak list with TF name
2) Features GRanges to intersect with
3) Count the number of intersections with a feature

Note that in the set up here we are calling functions. Functions are useful when performing fundemental operations (e.g. reading in the peak files and naming them by TF)

#### To make a function we use a markdown starting with: #'
The best practices for a function markdown are to include:
(i) a brief description 
(ii) @description: a more explicit description of what is being done.
(iii) @param: will define any parameters needed for the function (e.g. file path to peak files)
Here is an example of the markdown before coding the function:

#' import_peaks: .bed files as a list
#' 
#' @description 
#' this function will take consensus peak files and name them by the DNA binding protein and return a list
#' 
#' @param consensus_file_path the path to consensus peak files

# Now let's look at at a function to import peaks

```{r import_peaks}
# First we will set the function name and the following operations, in this case the function is called "import_peaks"
# The core principle is: 
# x <- function (param) {
# what you want done
# return(object created)
#}

import_peaks <- function(consensus_file_path = "/Shares/rinn_class/data/k562_chip/analysis/00_consensus_peaks/results/") {
  #setting object peak_files as a list of file names
  peak_files <- list.files(consensus_file_path, full.names = T)
  #this pulls out the TF names from the file names
  file_names <- str_extract(peak_files, "[\\w-]+\\.bed")
  tf_name <- str_extract(file_names, "^[^_]+(?=_)")
  
  # Now we want to make a list of GRanges objects called "peak_list" that is appended to the 
  peak_list <- c()
  for(i in 1:length(peak_files)) {
    # Import peaks
    peaks <- rtracklayer::import(peak_files[i])
    # Append this GRanges object to the of the list.
    peak_list <- c(peak_list, peaks)
    # Name the list elements by their TF name.
    names(peak_list)[length(peak_list)] <- tf_name[i]
  }
  # Here we want to return the output of the function
  # What do you think we want to return here? It would be written as return( varable_name)
  return(peak_list)
}


# Now that you've loaded the function, you still need to call the function.
# Do that here and assign the output to a variable.
peak_list <- import_peaks()

# Q: Do you need an argument for this function? HINT: What is the default?



# Now we have return a list of all the peaks for each DNA binding protein which is usally our starting point 
# Run this and see the contents of the returned GRanges
# This will be very useful in doing "find overlaps" with genomic features such as promoters described below

```

Above we have a GRanges object with out DNA binding name and all the corresponding consensus peaks
We can use the function below to compare these peaks with numerous genomic features (such as promoters)
First let's see the core of the function 
? What is the code to make the required param "features" ? What format does it need to be in?

```{r overlap with genomic features}
#' function finds overlaps between consensus_peaks and genomic features 
#' 
#' @description get_overlapping_peaks
#' this function will intersect the consesus_peaks with gene_body, promoters, mRNA_promoters, lncRNA_promoters, te_family
#' 
#' @param features
#'  set of genomic features as a GRanges object
#'  
#' @param peak_list
#' #list of peaks of dna binding proteins that will be intersected

# Great, so now we can make a function called "get_overlapping_peaks" that is re-usable for all features in features or just one feature such as a promoter :)
# Try running this function as a group to intersect your group genomic feature for class purpose -- let's all focus on "promoters" in Gencode 
## Hint you will need to make a "features" file that is compatable for findOverlaps with the output above. Double hint: GencodeV32.gtf
# ? What would "features" look like for promoters and what was the code you used to derive it



```

















# Ok so here is a solution to make a function that intersects all DNA binding proteins and all promoters

```{r}
#' function to subset features for promomters. 
#' 
#' @description feature_subset
#' Take a gencode gtf to subset the biotype of promoters we want as a set of GRanges
#' 
#' @param gencode_gr
#'  set of genomic features as a GRanges object
#'  
#' @param biotype
#' this takes "lncRNA" or "protein-coding" as input for promoter type
#'
#' @param upstream
#'To add upstream sequence to feature subset
#'
#' @param downstream
#'To add downstream sequence to feature subset

# First we are setting the "get_promoter_regions" function, but one param is up and downstream what should we use here?
get_promoter_regions <- function(gencode_gr, biotype, upstream = ?, downstream = ?) {
  
  #What would we want the ? to be?  Hint what is in the genocode GRanges so we could find overlaps with mRNA or lncRNA promoters
  genes <- gencode_gr[gencode_gr$type == "?"]
  genes <- genes[which(genes$gene_type == ?)]
  # What do we want here where the question mark is ? What will it include ?
  proms <- GenomicRanges::promoters(genes, upstream = ?, downstream = ?)

  return(proms)
}

```

Excellent we now have an intersection of all consensus peaks with mRNA and lncRNA promoters and a function that can be used for any subset of "features" in gencode GRanges.
Now let's make a function to count the number of times there is overlap of a given DNA binding protien (TF) and an element in features
Let's call this function "count_peaks_per_feature"

```{r}

#' count_peaks_per_feature: function to summarize the number of events in features on each individual promoter. 
#' 
#' @description 
#' Take a gencode gtf to subset the biotype of promoters we want as a set of GRanges
#' 
#' @param features
#' set of genomic features as a GRanges object
#'  
#' @param peak_list
#' #list of peaks of dna binding proteins that will be intersected

#? which is a list of GRanges and which is a GRanges object
count_peaks_per_feature <- function(features, peak_list) {

  peak_count <- matrix(numeric(), ncol = length(features), nrow = 0)
  
  for(j in 1:length(peak_list)) {
    ov <- countOverlaps(features, peak_list[[j]])
    peak_count <- rbind(peak_count, ov)
    rownames(peak_count)[nrow(peak_count)] <- names(peak_list)[j]
    colnames(peak_count) <- features$gene_id
  }
  return(peak_count)
}

```


```{r}
# debugging excercise within a function 

source("temp.R")
# prom_peaks <- count_peaks_per_feature_r(proms, peak_list)
  
```



Class excercise: use these functions to find the promoter with the most DNA binding proteins that "co-occur" on that promoter? 
Which mRNA and lncRNA promoter has the most DNA binding events?
>> Present this on wednesday for your group 
>> advanced use the function below to find the DNA binding protien that is most bound to the TE family or subset of your choosing.


```{r}
#Alternative hint
### Hint
peak_list <- import_peaks("/Shares/rinn_class/data/k562_chip/analysis/00_consensus_peaks/results/")
gencode_gr <- rtracklayer::import("/Shares/rinn_class/data/genomes/human/gencode/v32/gencode.v32.annotation.gtf")

# Here we are using the function "get_promoter_regions" to grab both the lncRNA and mRNA genes
promoters <- get_promoter_regions(gencode_gr, biotype = c("lncRNA", "protein_coding"))

# Possible ways to filter
promoter_peak_count <- count_peaks_per_feature(promoters, peak_list)

# Remove columns with no TF binding
promoter_peak_count <- promoter_peak_count[,which(colSums(promoter_peak_count) > 0)]

# Remove TFs with no promoter peaks
no_promoter_binding <- rownames(promoter_peak_count)[which(rowSums(promoter_peak_count) == 0)]
promoter_peak_count <- promoter_peak_count[which(rowSums(promoter_peak_count) > 0),]

# Make this binary -- make it a co-occurence matrix.
promoter_peak_occurence <- matrix(as.numeric(promoter_peak_count > 0), 
                           nrow = dim(promoter_peak_count)[1],
                           ncol = dim(promoter_peak_count)[2])
rownames(promoter_peak_occurence) <- rownames(promoter_peak_count)
colnames(promoter_peak_occurence) <- colnames(promoter_peak_count)

```


```{r}
# Advanced lesson for TE group

# subset rmsk

import_repeatmasker <- function(rmsk_file = "/Shares/rinn_class/data/k562_chip/util/rmsk.txt") {
  
  rmsk <- read.table(file = rmsk_file)
  colnames(rmsk) <- c("bin", "swScore", "milliDiv", "milliDel", "milliIns",
                      "genoName", "genoStart", "genoEnd", "genoLeft", "strand",
                      "repName", "repClass", "repFamily", "repStart",	"repEnd",
                      "repLeft",	"id")
  
  rmsk_gr <- GRanges(seqnames = rmsk$genoName,
                     IRanges(rmsk$genoStart, rmsk$genoEnd),
                     strand = rmsk$strand)
  
  # Add metadata colums
  rmsk_gr$rep_class <- rmsk$repClass
  rmsk_gr$rep_family <- rmsk$repFamily
  rmsk_gr$rep_name <- rmsk$repName
  
  return(rmsk_gr)
}

subset_rmsk <- function(rmsk_gr, rep_level = "family") {
  # rep_level needs to be one of "class", "family", or "name"
  if(!(rep_level %in% c("class", "family", "name"))) {
    stop("Repeat level needs to be either: class, family, or name.")
  } 
  
  level_column <- grep(rep_level, names(rmsk_gr@elementMetadata))
  rep_levels <- unique(rmsk_gr@elementMetadata[,level_column])
  
  rmsk_list <- c()
  for(i in 1:length(rep_levels)) {
    rmsk_list <- c(rmsk_list, list(rmsk_gr[rmsk_gr@elementMetadata[,level_column] == rep_levels[i]]))
    names(rmsk_list)[length(rmsk_list)] <- rep_levels[[i]]
  }
  return(rmsk_list)
}



```





