---
title: "Consensus Peak Set"
author: "JR & Michael"
date: "3/7/2020"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE)
library(GenomicRanges)
library(rtracklayer)
library(tidyverse)
library(ggpubr)
source("../util/_setup.R")
source("../util/intersect_functions.R")
```


# Purpose: Create consensus peak sets

To be conservative in our peak calls and be sure that we're dealing with replicatable peaks in the following analyses, here will will create a consensus peak .bed file for each DNA binding protein by taking only those peaks which overlap in all replicate experiments. 

The resulting peaks will consist of peaks which overlapped by at least on base pair in each replicate and will use the `GenomicRanges::reduce` function to merge the peaks by taking the outer boundaries of overlapping peaks. This strategy may widen some peaks, but will ensure that each peak in the resulting peak set has evidence in all experiments performed for that DNA binding protein. 


```{r create-consensus-peaks, warnings=FALSE, message=FALSE}
# For details on this function (create_concensus_peaks), see util/intersect_functions.R
# In this process we also filter to only the peaks on canonical chromosomes (e.g, non chr contigs).
consensus_peaks <- create_consensus_peaks(broadpeakfilepath = "/Shares/rinn_class/data/k562_chip/results/bwa/mergedLibrary/macs/broadPeak")
```

## Number of peaks threshold

Since we want to study robust DNA binding proteins in K562 cells, we will implement a cutoff for the minimum number of peaks a DBP must have in order to be considered in the following analyses. 

First, we need to filter out those DBPs that don't have any peaks common in replicates. 
```{r dist-num-peaks}
# This data frame will be the start of what we build into peak_occurence_df
# Plotting distributions of number of peaks that concur between replicates (note many @ 0)
num_peaks_df <- data.frame("dbp" = names(consensus_peaks),
                           "num_peaks" = sapply(consensus_peaks, length))
g <- ggplot(num_peaks_df, aes(x = num_peaks))
g + geom_histogram(bins = 70) +
  xlab("Number of consensus peaks") +
  ylab("Count") +
  ggtitle("Distribution of number of consensus peaks")
ggsave("figures/consensus_peaks_histogram.pdf")
```

#### DBPs that did not have a single peak in common in replicates

The proteins with zero consensus peaks are: `r paste(num_peaks_df[which(num_peaks_df$num_peaks == 0), "dbp"], collapse = " ")`

#### Peaks cutoff: now that we have replicate concordance, we next filter on having at least 250 peaks in common, based on distribution of 

#TODO what quartile is 250 peaks we should justify above.

```{r min-peaks-cutoff}
# We're going to apply a cutoff at 250 peaks           
num_peaks_threshold <- 250
threshold_percentile <- round(ecdf(num_peaks_df$num_peaks)(num_peaks_threshold)*100,1)

filtered_consensus_peaks <- consensus_peaks[num_peaks_df$num_peaks > num_peaks_threshold]
```

Since this captures the majority of DPBs and still provides a reasonable number of peaks to work with, we chose a cutoff of `r num_peaks_threshold` peaks. This corresponds to a cutoff at the `r threshold_percentile`th percentile and results in losing the following proteins: `r paste(num_peaks_df[which(num_peaks_df$num_peaks <= 250), "dbp"], collapse = " ")`


```{r export-peak-lists}
# Export the peak lists.
for(i in 1:length(filtered_consensus_peaks)) {
  rtracklayer::export(filtered_consensus_peaks[[i]], paste0("results/consensus_peaks/filtered_by_peaks/",
                                                   names(filtered_consensus_peaks)[i], 
                                                   "_consensus_peaks_filter.bed"))
}
```


## Summary of consensus peaks
Here we'll look at a few characteristics of DBPs that were consistent across replicates and have at least 250 binding events.

#### Goal: Investigate Peak width properties 

```{r total-peak-length}
# Subset to remaining peaks
num_peaks_df <- num_peaks_df %>% filter(dbp %in% names(filtered_consensus_peaks))

# Calculate the total peak width (bp bound by all peaks)
num_peaks_df$total_peak_length <- sapply(filtered_consensus_peaks, function(peaks) sum(width(peaks)))
write_csv(num_peaks_df, "results/number_of_peaks_per_dbp.csv")

g <- ggplot(num_peaks_df, aes(x = num_peaks, y = total_peak_length, label = dbp))
g + geom_point() + 
  geom_smooth(method = "lm", se = FALSE, color = "black", lty = 2) +
  geom_text() +
  ylab("BP covered") +
  xlab("Number of peaks") +
  ggtitle("Peak count vs. total bases covered")
ggsave("figures/peak_count_vs_peak_length.pdf")
```


#### eGFP peak count distribution

```{r egfp-peaks, warning=FALSE}
num_peaks_df$egfp_labeled <- FALSE
num_peaks_df[grep("eGFP", num_peaks_df$dbp), "egfp_labeled"] <- TRUE

g <- ggplot(num_peaks_df, aes(x = num_peaks, color = egfp_labeled))
g + stat_ecdf() + 
  scale_color_manual(values = c("#424242","#a8404c")) +
  ylab("Cumulative density") +
  xlab("Consensus peaks") + 
  ggtitle("eGFP vs. non-eGFP peak count")
  ggsave("figures/egfp_peak_vs_non_egfp_number_consensus_peaks.pdf")


# Test whether the distribution of peak counts is the same for eGFP labeled proteins
# Note: the ks-test will give a p-value estimate since of of the values are tied.
ks.test(x = num_peaks_df[which(num_peaks_df$egfp_labeled),"num_peaks"],
        y = num_peaks_df[which(!num_peaks_df$egfp_labeled),"num_peaks"])
```

There is not a difference in peak counts for eGFP labelled proteins.

#### Plotting the MEAN width of consensus peaks for each DBP

```{r peak-width-distribution}
# Let's grab the peak widths for each DBP
peak_widths_df <- lapply(filtered_consensus_peaks, 
                      function(peaks) paste(width(peaks), collapse = ";")) %>%
  as.data.frame() %>%
  pivot_longer(cols = everything(),
               names_to = "dbp",
               values_to = "peak_width") %>%
  separate_rows("peak_width", sep = ";", convert = T)

# Summary data frame (math applied in socket settings). This is set up to encapsulate mean, median, sd, cv of peak widths for each DBP
peak_widths_summary <- peak_widths_df %>% 
  group_by(dbp) %>%
  summarize("mean_width" = mean(peak_width),
            "median_width" = median(peak_width),
            "sd_width" = sd(peak_width),
            "cv_width" = (sd_width/mean_width)*100)

# Plotting resulting peak_width_summary distribution of consensus peaks for each DBP. 
#NOTE: many have 0 consensus peaks

g <- ggplot(peak_widths_summary, aes(x = mean_width))
g + geom_histogram(bins = 60) + 
  xlab("Mean peak width") +
  ylab("Count") +
  ggtitle("Peak width distribution")
ggsave("figures/peak_width_distribution.pdf")

# Plotting the mean width versus the covariance (cv) of peak widths for that DBP. This will distinguish if a DBP has a uniform consensus peak width or high variance in consensus peak width.
g <- ggplot(peak_widths_summary, aes(x = mean_width, y = cv_width, label = dbp))
g + geom_point() + 
  geom_text() + 
  xlab("Mean peak width") +
  ylab("CV peak width") +
  ggtitle("Peak width: mean vs. coefficient of variation")
ggsave("figures/peak_width_vs_cv.pdf")
```



It seems that most DBPs have a relatively small peak width ~ 1000bp +/- 500bp and relatively small CV ~100bp. However, a few including POLII subunits have wider peaks widths. RFX1 is unique in that it has a typical mean peak width, but a very large CV. Indicating that while most peaks are in the typical range, it may have some very large peaks. 

```{r polII-peak-widths}
# Let's look at PolII and SUPT5H: The 'core' transcriptional machinery in particular
g <- ggplot(peak_widths_df %>% filter(dbp %in% c("POLR2A", "POLR2B", "SUPT5H")), aes(x = log10(peak_width)))
g + geom_histogram(bins = 100) + 
  facet_grid(dbp~., scales = "free_y") + 
  xlab("log10(Peak width)") +
  ylab("Count") +
  ggtitle("Peak width distribution: POLII")
ggsave("figures/peak_width_distribution_polII.pdf")
ggsave("figures/peak_width_distribution_polII.png")
```

So interestingly it seems that there are two peaks in POLR2A's binding, but not for the other subunits. This may indicate two binding modes for POLR2A.



```{r rfx1-peak-widths}
# Here is a more extreme example that has a high variance in peak widths (RFX1) 
g <- ggplot(peak_widths_df %>% filter(dbp %in% c("RFX1")), aes(x = log10(peak_width)))
g + geom_histogram(bins = 100) + 
  facet_grid(dbp~., scales = "free_y") + 
  xlab("log10(Peak width)") +
  ylab("Count") +
  ggtitle("Peak width distribution: RFX1")
ggsave("figures/peak_width_distribution_RFX1.pdf")


```

So the extremely long peak widths from RFX1 seem to be outliers. There are only `r nrow(peak_widths_df %>% filter(dbp == "RFX1", peak_width > 10^4))` peaks greater than 10kb. Let's look at an example anyway.

```{r}
library(Gviz)

library(GenomicRanges)
data(cpgIslands)
class(cpgIslands)
## [1] "GRanges"
## attr(,"package")
## [1] "GenomicRanges"
chr <- as.character(unique(seqnames(cpgIslands)))
gen <- genome(cpgIslands)
atrack <- AnnotationTrack(cpgIslands, name = "CpG")
gtrack <- GenomeAxisTrack()
itrack <- IdeogramTrack(genome = gen, chromosome = chr)
data(geneModels)
grtrack <- GeneRegionTrack(geneModels, genome = gen,
                           chromosome = "chr7", name = "Gene Model")
plotTracks(list(itrack, gtrack, atrack, grtrack))


gen <- "hg38"

names(geneModels)
# Let's make a gencode "genemodel"
gencode_gm <- gencode_gr %>%
  as.data.frame() %>%
  filter(model == "exon") %>%
  dplyr::select(seqnames, start, end, width, strand, gene_type,
                gene_id, exon_id, transcript_id, gene_name) 

names(gencode_gm) <- c("chromosome", "start", "end", "width", "strand",
                       "feature", "gene", "exon", "transcript", "symbol")

reg_gm <- gencode_gm %>% filter(chromosome == "chr20",
                                start >= 53609523,
                                end <= 53896205)
itrack <- IdeogramTrack(genome = "hg38", chromosome = "chr20")
grtrack <- GeneRegionTrack(reg_gm, genome = "hg38",
                           chromosome = "chr20", name = "Gencode genes",
                           showId = TRUE,
                           fill = "gray70")
?GeneRegionTrack
plotTracks(list(itrack, gtrack, grtrack))

# Read in RFX1 peaks. 
replicate_peak_files <- list.files("/Shares/rinn_class/data/k562_chip/results/bwa/mergedLibrary/macs/broadPeak/", full.names = T)
rfx1_replicate_peak_files <- replicate_peak_files[grep("RFX1", replicate_peak_files)]
rfx1_replicate_peak_files <- rfx1_replicate_peak_files[grep("peaks.broadPeak", rfx1_replicate_peak_files)]

gr <- GRanges(seqnames = "chr20", IRanges(53609523, 53896205), strand = "*")

pf1 <- lapply(rfx1_replicate_peak_files, read_peaks, filter_to_canonical_chr = T)
# filter to just canonical chromosomes. 
reg_pf1 <- subsetByOverlaps(pf1[[1]], gr)
reg_pf2 <- subsetByOverlaps(pf1[[2]], gr)
reg_pf3 <- subsetByOverlaps(pf1[[3]], gr)
reg_pf4 <- subsetByOverlaps(pf1[[4]], gr)
atrack1 <- AnnotationTrack(reg_pf1, col = "#424242", fill = "#424242", name = "Peaks rep1")
atrack2 <- AnnotationTrack(reg_pf2, col = "#424242", fill = "#424242", name = "Peaks rep2")
atrack3 <- AnnotationTrack(reg_pf3, col = "#424242", fill = "#424242", name = "Peaks rep3")
atrack4 <- AnnotationTrack(reg_pf4, col = "#424242", fill = "#424242", name = "Peaks rep4")
plotTracks(list(itrack, gtrack, atrack1, atrack2, atrack3, atrack4, grtrack))

# K let's get some bigwigs in here. 
bws <- list.files("/Shares/rinn_class/data/k562_chip/results/bwa/mergedLibrary/bigwig/", full.names = T)
rfx1_bigwigs <- bws[grep("RFX1", bws)]
rfx1_bigwigs

dTrack1 <- DataTrack(range = rfx1_bigwigs[[1]], genome = "hg38", type = "h", 
                     chromosome = "chr20", name = "RFX1 rep1", ylim = c(0,0.6),
                     size = 1, col = "#212121")
dTrack2 <- DataTrack(range = rfx1_bigwigs[[2]], genome = "hg38", type = "h", 
                     chromosome = "chr20", name = "RFX1 rep2", ylim = c(0,0.6),
                     size = 1, col = "#212121")
dTrack3 <- DataTrack(range = rfx1_bigwigs[[3]], genome = "hg38", type = "h", 
                     chromosome = "chr20", name = "RFX1 rep3", ylim = c(0,0.6),
                     size = 1, col = "#212121")
dTrack4 <- DataTrack(range = rfx1_bigwigs[[4]], genome = "hg38", type = "h", 
                     chromosome = "chr20", name = "RFX1 rep4", ylim = c(0,0.6),
                     size = 1, col = "#212121")
plotTracks(list(itrack, gtrack, atrack1, dTrack1, atrack2, 
                dTrack2, atrack3, 
                dTrack3, atrack4,
                dTrack4, grtrack),
           background.title = "white",
           fontcolor = "black",
           rotation.title = 90,
           cex.title = 0.7,
           col.title = "black",
           col.axis = "black", 
           main = "RFX1 largest peak",
           cex.main = 0.8)
availableDisplayPars("GdObject")
availableDisplayPars("DataTrack")
?DataTrack
# Let's grab the bigwig files
library(ggbio)
p.ideo <- Ideogram(genome = "hg38")
p.ideo
p.ideo + xlim(GRanges("chr2", IRanges(2e8, 2e8+10000000)))

library(biovizBase)
library(Homo.sapiens)
class(Homo.sapiens)
data(genesymbol, package = "biovizBase")
wh <- genesymbol[c("BRCA1", "NBR1")]
wh <- range(wh, ignore.strand = TRUE)
p.txdb <- autoplot(Homo.sapiens, which = wh)
p.txdb


library(TxDb.Hsapiens.UCSC.hg38.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene
gr.txdb <- crunch(txdb, which = wh)
## change column to 'model'
colnames(values(gr.txdb))[4] <- "model"
grl <- split(gr.txdb, gr.txdb$tx_id)
## fake some randome names
names(grl) <- sample(LETTERS, size = length(grl), replace = TRUE)
grl

gencode_gr <- rtracklayer::import("../../../genomes/human/gencode/v32/gencode.v32.annotation.gtf")

colnames(values(gencode_gr))[2] <- "model"
gencode_gr_no_genes <- gencode_gr[gencode_gr$model != "gene"]
gencode_grl <- split(gencode_gr, gencode_gr$transcript_id)

# autoplot(grl, aes(type = model))
chr20
53609523
library(AnnotationFilter)
53896205

gr <- GRanges(seqnames = "chr20", IRanges(53609523, 53896205), strand = "*")
gr_subset <- subsetByOverlaps(gencode_grl, gr)

autoplot(gr_subset, names.expr = "gene_name", label = TRUE)
?autoplot
library('GenomicFeatures')
txdb = makeTxDbFromGFF('../../../genomes/human/gencode/v32/gencode.v32.annotation.gtf')

library(AnnotationDbi)
saveDb(txdb, 'results/gencode.v32.annotation.sqlite')
wh <- genesymbol[c("BRCA1", "NBR1")]
wh <- range(wh, ignore.strand = TRUE)

library(AnnotationDbi)
txdb <- loadDb(file = 'results/gencode.v32.annotation.sqlite')

txdb
# autoplot(txdb, which

autoplot(txdb, which = gr, names.expr = "gene_name")
# gr.txdb <- crunch(txdb, which = wh)
# ## change column to 'model'
# colnames(values(gr.txdb))[4] <- "model"
# grl <- split(gr.txdb, gr.txdb$tx_id)
# ## fake some randome names
# names(grl) <- sample(LETTERS, size = length(grl), replace = TRUE)
# grl

bws <- list.files("/Shares/rinn_class/data/k562_chip/results/bwa/mergedLibrary/bigwig/", full.names = T)
rfx1_bigwigs <- bws[grep("RFX1", bws)]
rfx1_bigwigs
input_bigwigs <- bws[grep("input", bws)]
# Remove egfp controls
input_bigwigs <- input_bigwigs[!grepl("eGFP", input_bigwigs)]
replicate_peak_files <- list.files("/Shares/rinn_class/data/k562_chip/results/bwa/mergedLibrary/macs/broadPeak/")
rfx1_replicate_peak_files <- replicate_peak_files[grep("RFX1", replicate_peak_files)]
rfx1_replicate_peak_files <- rfx1_replicate_peak_files[grep("broadPeak", rfx1_replicate_peak_files)]

replicate_peak_files


#RFX1 has most peaks around 1kb but a long tail of much larger peaks maxing out at 286,683 ~287KB!
max(peak_widths_df[which(peak_widths_df$dbp == "RFX1" & peak_widths_df$peak_width > 3000),"peak_width"])

#TODO browser example of large peak window?
```

### Example of RFX1 peak

```{r}
library(wiggleplotr)

rfx1_samples <- read_csv("../../samplesheet.csv") 



rfx1_peaks <- filtered_consensus_peaks[["RFX1"]] %>%
  as.data.frame() %>%
  arrange(-width)
?plotTranscripts
```

And it does seem that RFX1 has just `r nrow(peak_widths_df[which(peak_widths_df$dbp == "RFX1" & peak_widths_df$peak_width > 3000),])` peaks which are above 3000 bps which is skewing the distribution. With one peak that has a width of `r max(peak_widths_df[which(peak_widths_df$dbp == "RFX1" & peak_widths_df$peak_width > 3000),"peak_width"])`.


#### DNA binding protein distribution at promoters

The purpose of this code is to determine the distribution of DNA-binding proteins (DBPs) at promoters. We will look at the number of DBPs bound at a given promoter as well as whether there are any promoters which are not bound by any DBPs. We will also ask whether any DBPs do not bind to any promoters.

```{r gencode-import}
gencode_gr <- rtracklayer::import("/Shares/rinn_class/data/genomes/human/gencode/v32/gencode.v32.annotation.gtf")
# For convenience, we'll export these lists to read in the following analyses. 
# We are using the function 'get_promoter_regions' see "../util/intersect_functions.R"
lncrna_mrna_promoters <- get_promoter_regions(gencode_gr, biotype = c("lncRNA", "protein_coding"))
rtracklayer::export(lncrna_mrna_promoters, "results/lncrna_mrna_promoters.gtf")

lncrna_promoters <- get_promoter_regions(gencode_gr, biotype = "lncRNA")
rtracklayer::export(lncrna_promoters, "results/lncrna_promoters.gtf")

mrna_promoters <- get_promoter_regions(gencode_gr, biotype = "protein_coding")
rtracklayer::export(mrna_promoters, "results/mrna_promoters.gtf")
```


# Promoter overlaps compared to toal number of peaks for each DBP
# Goal: first to count the number of DBPs on each promoter (this uses 'count_peaks_per_feature' function "../util/intersect_funcitons.R")

```{r}
num_peaks_df <- data.frame("dbp" = names(filtered_consensus_peaks),
                           "num_peaks" = sapply(filtered_consensus_peaks, length))
# Getting number of overlaps of DBPs per promoter(type)
promoter_peak_counts <- count_peaks_per_feature(lncrna_mrna_promoters, filtered_consensus_peaks, type = "counts")

# Counting up overlaps for all promoters and lncRNA and mRNA seperately.
num_peaks_df$peaks_overlapping_promoters <- rowSums(promoter_peak_counts)
num_peaks_df$peaks_overlapping_lncrna_promoters <- rowSums(promoter_peak_counts[,lncrna_promoters$gene_id])
num_peaks_df$peaks_overlapping_mrna_promoters <- rowSums(promoter_peak_counts[,mrna_promoters$gene_id])

# Plotting the number of peaks per DBP and how many promoters were overlapped.
ggplot(num_peaks_df,
       aes(x = num_peaks, y = peaks_overlapping_promoters)) +
  scale_color_manual(values = c("#424242", "#a8404c"))+
  xlab("Peaks per DBP") +
  ylab("Number of peaks overlapping promoters") +
  ggtitle("Relationship Between Number of DBP Peaks and Promoter Overlaps")+
  geom_point() +
  geom_abline(slope = 1, linetype="dashed") +
  geom_smooth(method = "lm", se=FALSE) +
  stat_regline_equation(label.x = 35000, label.y = 16000) + 
  ylim(0,60100) +
  xlim(0,60100)

ggsave("figures/peak_number_vs_promoter_overlap.pdf")
```


# Goal: let's bring all of these data together into a dataframe, that we will keep adding columns to as we go (rows are promoters).

```{r count-peaks-per-promoter}

# First we will establish the promoter peak occurence matrix (rows = promoters, cols = dbp binding {0,1})
promoter_peak_occurence <- count_peaks_per_feature(lncrna_mrna_promoters, filtered_consensus_peaks, 
                                               type = "occurrence")
# Output to promoter_peak_occurecne_matrix
write.table(promoter_peak_occurence, "results/lncrna_mrna_promoter_peak_occurence_matrix.tsv")

# Now we want to make into a data frame using the promoter annotations as rows and attributes as columns.
# We will use lncrna_mrna_promoters to index "all promoters"
# First make sure promoter_peak_occurence and lncrna_mrna_promoters are in the same order
stopifnot(all(colnames(promoter_peak_occurence) == lncrna_mrna_promoters$gene_id))

# Make a data frame of binding events per promoter.
peak_occurence_df <- data.frame("gene_id" = colnames(promoter_peak_occurence),
                                "gene_name" = lncrna_mrna_promoters$gene_name,
                                "gene_type" = lncrna_mrna_promoters$gene_type,
                                "3kb_up_tss_start" = lncrna_mrna_promoters@ranges@start,
                                "strand" = lncrna_mrna_promoters@strand,
                                "number_of_dbp" = colSums(promoter_peak_occurence))

# This is the CSV file we will start building upon adding columns of properties as we analyze them
# The output file name will change based on what is added later, but the "peak_occurence_df" will be used throughout.
write_csv(peak_occurence_df, "results/peak_occurence_dataframe.csv")

```


# Now that we have our consensus peaks organized, characterized and filtered above we return to our goal
# Goal: density plot the number of DBPs per promoter 
```{r promoter-peak-distribution}
# Make a density plot of all promoter binding events per goal #1.
g <- ggplot(peak_occurence_df, aes(x = number_of_dbp))
g + geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
  theme_paperwhite() +
  xlab(expression("Number of DBPs")) +
  ylab(expression("Density")) +
  ggtitle("Promoter binding events",
          subtitle = "mRNA and lncRNA genes")
ggsave("figures/k562_promoter_binding_density.png")
ggsave("figures/k562_promoter_binding_density.pdf")

g <- ggplot(peak_occurence_df, aes(x = number_of_dbp))
g + geom_histogram(color = "#424242", fill = "#424242") +
  theme_paperwhite() +
  xlab(expression("Number of DBPs")) +
  ylab(expression("Count")) +
  ggtitle("Promoter binding events",
          subtitle = "mRNA and lncRNA genes")
ggsave("figures/k562_promoter_binding_histogram.png")
ggsave("figures/k562_promoter_binding_histogram.pdf")
```

#TODO : Print, min, max, range, deriviative around 15? Kinda cool result to include



```{r unbound-promoters}
# just locating the unbound promoters in any experiment
unbound_promoters <- peak_occurence_df %>% filter(peak_occurence_df$number_of_dbp < 1)
write_csv(unbound_promoters, "results/unbound_lncrna_mrna_promoters.csv")
```

```{r}
# Determining how many DBPs don't bind to any promoters.
dbp_promoter_ovl <- get_overlapping_peaks(lncrna_mrna_promoters, filtered_consensus_peaks)
num_ovl <- sapply(dbp_promoter_ovl, length)
```

There are no proteins that don't overlap any promoters and the least number of promoters bound is `r names(num_ovl)[which(num_ovl == min(num_ovl))]` which overlapped `r min(num_ovl)`

```{r}
##TODO Refector or remove?

promoter_peak_counts <- count_peaks_per_feature(lncrna_mrna_promoters, filtered_consensus_peaks,                                               type = "counts")

num_peaks_df$peaks_overlapping_lncrna_promoters <- rowSums(promoter_peak_counts[, lncrna_mrna_promoters$gene_id])
num_peaks_df$peaks_overlapping_mrna_promoters <- rowSums(promoter_peak_counts[,mrna_promoters$gene_id])
num_peaks_dfl <- num_peaks_df %>% pivot_longer(4:5, names_to = "promoter_type",
                                               values_to = "peaks_overlapping_promoters")

ggplot(num_peaks_dfl,
       aes(x = num_peaks, y = peaks_overlapping_promoters, color = promoter_type)) +
  scale_color_manual(values = c("#424242", "#a8404c"))+
  xlab("Peaks per TF") +
  ylab("Peaks Overlapping Promoters") +
  ggtitle("Relationship Between Number of TF Peaks and Promoter Overlaps")+
  geom_point() +
  geom_abline(slope = 1, linetype="dashed") +
  geom_smooth(method = "lm", se=FALSE)
ggsave("figures/peaks_overlaps_relationship.png", height = 5, width = 8)
ggsave("figures/peaks_overlaps_relationship.pdf", height = 5, width = 8)
```




