---
title: "DNA-binding at promoters"
output: html_document
editor_options:
  chunk_output_type: console
---


```{r setup, include=FALSE}
options(stringsAsFactors = F)
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
library(GenomicFeatures)
library(rtracklayer)
library(ggrepel)
# BiocManager::install("Rsubread", version = 3.8)
library(Rsubread)
```

The purpose of this code is to determine the distribution of DNA-binding proteins (DBPs) at promoters. We will look at the number of DBPs bound at a given promoter as well as whether there are any promoters which are not bound by any DBPs. We will also ask whether any DBPs do not bind to any promoters.


Here we are going to make density plots for the number of DBPs (DNA-binding proteins) at a given promoter.
```{r}
#get functions needed for this code
source("../util/intersect_functions.R")

#Read in consensus peaks as a list.
#import_peaks is a function
peak_list <- import_peaks()

#Read in gencode annotated genome
gtf <- rtracklayer::import("/Shares/rinn_class/data/genomes/human/gencode/v32/gencode.v32.annotation.gtf")

#Get the promoter regions of each gene.
#get_promoter_regions is a function
proms <- get_promoter_regions(gtf, biotype = c("lncRNA", "protein_coding"))

#Count peaks over promoters
#count_peaks_per_feature is a function
promoter_peak_count <- count_peaks_per_feature(proms, peak_list)
```

```{r}
# Make promoter_peak_count binary -- a co-occurence matrix.
promoter_peak_occurence <- matrix(as.numeric(promoter_peak_count > 0),
                           nrow = dim(promoter_peak_count)[1],
                           ncol = dim(promoter_peak_count)[2])
rownames(promoter_peak_occurence) <- rownames(promoter_peak_count)
colnames(promoter_peak_occurence) <- colnames(promoter_peak_count)

# Make a data frame of binding events per promoter.
peak_occurence_df <- data.frame("gene_id" = colnames(promoter_peak_occurence),
                                "number_of_tfs" = colSums(promoter_peak_occurence))

```


```{r}
# Make a density plot of all promoter binding events per goal #1.
g <- ggplot(peak_occurence_df, aes(x = number_of_tfs))
g + geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
  theme_bw() +
  xlab(expression("Number of TFs")) +
  ylab(expression("Density")) +
  ggtitle("Promoter binding events at all genes")
```
Figure 1. Density plot showing the number of binding events at a given promoter.

```{r}
ggsave("k562_promoter_binding_density.png")
ggsave("k562_promoter_binding_density.pdf")
```


Here we are going to determine if there are any promoters that are not bound by any DBPs.
```{r}
#set x to 1 
x <- 1
#filter peak_occurence_df for promoters with less than 1 TF bound
not_bound_proms <- filter(peak_occurence_df, peak_occurence_df$number_of_tfs < x)

#see not_bound_proms for list of promoters that are not bound
```

Here we are going to determine if any DBPs did not overlap with any promoters.
```{r}
#get_overlapping_peaks is a function
tf_promoter_ovl <- get_overlapping_peaks(proms, peak_list)

#get list of lengths of each GRanges object in tf_promoter_ovl
num_ovl <- sapply(tf_promoter_ovl, length)
#are any of these lengths zero? 
length(which(num_ovl == 0))

```

