---
title: "Clustering by Features"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
options(stringsAsFactors = FALSE)
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
source("/Shares/rinn_class/data/k562_chip/analysis/util/intersect_functions.R")
source("/Shares/rinn_class/data/k562_chip/analysis/util/_setup.R")

```


# The first step is to get a matrix of features by TFs.

```{r}
# We are going to start out with a function to import files called "import_peaks"
<<<<<<< HEAD
# This the filtered peak_list; final list of 161 TFs after filtering
=======
# this is the filtered peak_list; final list of 161 TFs after filtering
>>>>>>> 8f2e6f58d0ac5b922c7a589be865197fc4de85ae
peak_list <- import_peaks("/Shares/rinn_class/data/CLASS_2020/analysis/01_consensus_peaks/results/")

# Making the features (gencode) GRanges that has lots of features annotated
gencode_gr <- rtracklayer::import("/Shares/rinn_class/data/genomes/human/gencode/v32/gencode.v32.annotation.gtf")

<<<<<<< HEAD
# Here we are using the function "get_promoter_regions" to grab both the lncRNA and mRNA gene # promoters
=======
# Here we are using the function "get_promoter_regions" to grab both the lncRNA and mRNA promoters
>>>>>>> 8f2e6f58d0ac5b922c7a589be865197fc4de85ae
proms <- get_promoter_regions(gencode_gr, biotype = c("lncRNA", "protein_coding"))

# Counts the number of promoters that each TF binds to
promoter_peak_count <- count_peaks_per_feature(proms, peak_list)
<<<<<<< HEAD


=======
>>>>>>> 8f2e6f58d0ac5b922c7a589be865197fc4de85ae
```

```{r filtering}
# Now we want to remove any columns (promoters) where a promoter has no TF binding.
promoter_no_peak <- promoter_peak_count[,which(colSums(promoter_peak_count) < 1)]

### TFs bind to these promoters
promoter_positive_peaks <- promoter_peak_count[,which(colSums(promoter_peak_count) > 0)]
<<<<<<< HEAD

#yes_binding <- colnames(promoter_positive_peaks)

# Now we want to remove any columns (tfs) where a TF has bound no promoters
# no_binding <- colnames(promoter_no_peak)
=======

# Now we want to remove any columns (tfs) where a TF has bound no promoters
# Q: Which TFs bound no promoters?
no_binding <- colnames(promoter_no_peak)
### This is also a paper result!
>>>>>>> 8f2e6f58d0ac5b922c7a589be865197fc4de85ae

promoter_positive_peaks <- promoter_positive_peaks[which(rowSums(promoter_positive_peaks) > 0),]

# We're going to make this binary -- make it a co-occurence matrix where any number of
# peaks in a promoter for a tf is considered "binding" and so we're going to change all the
# values to zeros and ones.
promoter_bound <- promoter_peak_count > 0
## Make this into a numeric vector (0 or 1) instead of a boolean vector
promoter_occurence <- as.numeric(promoter_bound)

# We make this into a matrix with the original dimensions
promoter_peak_occurence <- matrix(promoter_occurence, 
                           nrow = dim(promoter_peak_count)[1],
                           ncol = dim(promoter_peak_count)[2])
# Let's transfer the row and column names
rownames(promoter_peak_occurence) <- rownames(promoter_peak_count)
colnames(promoter_peak_occurence) <- colnames(promoter_peak_count)
```

```{r}
<<<<<<< HEAD
# Find the distribution of DNA binding events across promoters that had at least 1 event
=======
# What is the distribution of DNA binding events across promoters that had at least 1 event?
>>>>>>> 8f2e6f58d0ac5b922c7a589be865197fc4de85ae

hist(promoter_positive_peaks)
```

# Now that we have a binary matrix and counts of number of features that intersect with DNA binding events -- let's cluster to see if the grouping makes sense
# We are using a binary matrix that calculates the distance betweeen each sample using the binary distance metric. 

```{r}

# Calculate a distance measure between the rows.
bin_hier <- hclust(dist(promoter_peak_occurence, method = "binary"))
plot(bin_hier)

# load library for ggdendro for the best trees we could find 
library(ggdendro)

ggdendro::ggdendrogram(bin_hier, rotate = TRUE,  size = 3)
ggsave("figures/hclust_binary_dist.pdf", height = 26, width = 6)

```








