---
title: "Clustering by Features"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
options(stringsAsFactors = FALSE)
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(ggdendro)
library(GenomicRanges)
library(umap)
library(dbscan)
library(ggrepel)
source("../util/intersect_functions.R")
source("../util/_setup.R")
```

# Global clustering

```{r import-data}
# We are going to start out with a function to import files called "import_peaks"
# This the filtered consensus_peaks; final list of 161 TFs after filtering
# consensus_peaks <- import_peaks("../01_consensus_peaks/results/consensus_peaks")

# Let's grab both the lncRNA and mRNA promoters
# proms <- rtracklayer::import("../01_consensus_peaks/results/lncrna_mrna_promoters.gtf")
lncrna_promoters <- rtracklayer::import("../01_consensus_peaks/results/lncrna_promoters.gtf")
mrna_promoters <- rtracklayer::import("../01_consensus_peaks/results/mrna_promoters.gtf")

# Counts the number of promoters that each TF binds to
promoter_peak_occurence <- read.table("../01_consensus_peaks/results/lncrna_mrna_promoter_peak_occurence_matrix.tsv") %>%
  as.matrix()
```


Now that we have a binary matrix and counts of number of features that intersect with DNA binding events -- let's cluster to see if the grouping makes sense. We are using a binary matrix that calculates the distance betweeen each sample using the binary distance metric. 

```{r hclust-all, fig.height=25, fig.width=5}
# Hierarchical clustering with binary distance measure
bin_hier <- hclust(dist(promoter_peak_occurence, method = "binary"))

ggdendro::ggdendrogram(bin_hier, rotate = TRUE,  size = 3)
ggsave("figures/all_hclust_binary_dist.pdf", height = 26, width = 6)
ggsave("figures/all_hclust_binary_dist.png", height = 26, width = 6)
```

## lncRNA promoter binding clustering

```{r hclust-lnrna, fig.height=25, fig.width=5}
lncrna_peak_occurence <- promoter_peak_occurence[,lncrna_promoters$gene_id]

bin_hier <- hclust(dist(lncrna_peak_occurence, method = "binary"))

ggdendro::ggdendrogram(bin_hier, rotate = TRUE,  size = 3)
ggsave("figures/lncrna_hclust_binary_dist.pdf", height = 26, width = 6)
ggsave("figures/lncrna_hclust_binary_dist.png", height = 26, width = 6)
```

## mRNA promoter binding clustering

```{r hclust-mrna, fig.height=25, fig.width=5}
mrna_peak_occurence <- promoter_peak_occurence[,mrna_promoters$gene_id]

bin_hier <- hclust(dist(lncrna_peak_occurence, method = "binary"))

ggdendro::ggdendrogram(bin_hier, rotate = TRUE,  size = 3)
ggsave("figures/mrna_hclust_binary_dist.pdf", height = 26, width = 6)
ggsave("figures/mrna_hclust_binary_dist.png", height = 26, width = 6)
```

# UMAP dimensionality reduction

Here we're going to perform UMAP and cluster with HDBscan to get a global view of DBP binding on lncRNA and mRNA promoters.

```{r umap-all, fig.height=10, fig.width=10.5, message=FALSE}
umap_params <- umap.defaults
umap_params$n_neighbors <- 4
umap_params$random_state <- 7748

umap_fit <- umap(promoter_peak_occurence, config = umap_params)
umap_df <- umap_fit$layout %>% as.data.frame() %>%
  rownames_to_column("dbp")

labels <- hdbscan(umap_fit$layout, minPts = 6)
umap_df$cluster <- factor(labels$cluster)
write_csv(umap_df, "results/umap_lncrna_mrna_promoters.csv")


col_pal <- c("#653EB3", "#B3823E", "#78B33E", "#3E9FB3", "#666666", "#fd8d3c", "#B33E52")
g <- ggplot(umap_df, aes(x = V1, y = V2, label = dbp, color = cluster))
g + geom_point() + 
  geom_text(data = umap_df %>% filter(dbp %in% c("POLR2A",
                                                  "POLR2B",
                                                  "SUPT5H",
                                                  "eGFP-POLR2H"))) + 
  ggtitle("UMAP: DBP promoter occupancy",
          subtitle = "mRNA and lncRNA promoters") +
  theme_paperwhite() +
  theme(axis.title = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank()) +
  scale_color_manual(values = col_pal)
ggsave("figures/umap_mrna_lncrna_promoters.png")
ggsave("figures/umap_mrna_lncrna_promoters.pdf")


# Let's make a df of the clusters for naming and gene ontology
clusters_df <- umap_df %>%
  dplyr::select(cluster, dbp) %>%
  group_by(cluster) %>%
  summarize(dbp = paste(dbp, collapse = " ")) %>%
  mutate(dbp = gsub("eGFP-","", dbp))
write_csv(clusters_df, "results/umap_clusters.csv")
knitr::kable(clusters_df)
```

### Sub clustering of largest cluster

```{r umap-large-cluster, fig.height=10, fig.width=10.5}
dbps_in_large_cluster <- umap_df[umap_df$cluster == 7, "dbp"]
lc_peak_occurence <- promoter_peak_occurence[dbps_in_large_cluster,]

umap_params <- umap.defaults
umap_params$n_neighbors <- 4
umap_params$min_dist <- 0.1
umap_params$spread <- 0.15
umap_params$random_state <- 2187

umap_fit <- umap(lc_peak_occurence, config = umap_params)
umap_sub <- umap_fit$layout %>% as.data.frame() %>%
  rownames_to_column("dbp")

labels <- hdbscan(umap_fit$layout, minPts = 4)
umap_sub$cluster <- factor(labels$cluster)
write_csv(umap_sub, "results/umap_largest_subcluster.csv")


col_pal <- c("#666666", "#9e9ac8", "#6b6ecf", "#B3823E", "#fd8d3c", "#78B33E", "#653EB3", "#6baed6", "#d6616b", "#3E9FB3")
g <- ggplot(umap_sub, aes(x = V1, y = V2, label = dbp, color = cluster))
g + geom_point()  +
    ggtitle("UMAP: large subcluster",
          subtitle = "mRNA and lncRNA promoters") +
  theme_paperwhite() +
  theme(axis.title = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank()) +
  scale_color_manual(values = col_pal) + 
  geom_text_repel()
ggsave("figures/umap_large_subcluster_promoters.png", height = 11, width = 11.5)
ggsave("figures/umap_large_subcluster_promoters.pdf", height = 11, width = 11.5)

clusters_df <- umap_sub %>%
  dplyr::select(cluster, dbp) %>%
  group_by(cluster) %>%
  summarize(dbp = paste(dbp, collapse = " ")) %>%
  mutate(dbp = gsub("eGFP-","", dbp))
write_csv(clusters_df, "results/umap_clusters.csv")
knitr::kable(clusters_df)
```


