---
title: "DNA-binding versus expression at promoters"
output: html_document
editor_options:
  chunk_output_type: console
---


```{r setup, include=FALSE}
options(stringsAsFactors = F)
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
library(GenomicFeatures)
library(rtracklayer)
#library(ggbio)
library(ggrepel)
library(Rsubread)
library(effectsize)
source("/Shares/rinn_class/data/k562_chip/analysis/util/intersect_functions.R")
```


#### This is where the data came from

total RNA-seq for K562 dataset: https://www.encodeproject.org/experiments/ENCSR885DVH/

# system("wget https://www.encodeproject.org/files/ENCFF625ZBS/@@download/ENCFF625ZBS.bam",
#        intern = FALSE, ignore.stdout = TRUE, ignore.stderr = TRUE)
# system("wget https://www.encodeproject.org/files/ENCFF630HEX/@@download/ENCFF630HEX.bam",
#        intern = FALSE, ignore.stdout = TRUE, ignore.stderr = TRUE)


```{r converting read counts to TPM}

# First we will take the bam file and find the number of reads that align over mRNA and lncRNA gene-bodies 

counts <- featureCounts(c("/Shares/rinn_class/data/k562_chip/analysis/09_expression/ENCFF625ZBS.bam",
                          "/Shares/rinn_class/data/k562_chip/analysis/09_expression/ENCFF630HEX.bam"),
                        annot.ext = "/Shares/rinn_class/data/genomes/human/gencode/v32/gencode.v32.annotation.gtf",
                        isGTFAnnotationFile = T,
                        isPairedEnd = TRUE,
                        nthreads = 16)

# First we will find the number of reads per kilobases of a given gene body (rpkm).

rpk <- counts$counts / (counts$annotation$Length/1000)
expression <- data.frame("rpk" = rpk) %>%
  rownames_to_column("gene_id") %>%
  pivot_longer(2:3, names_to = "sample", values_to = "rpk")

# Here we will convert rpkm to transcripts per million of reads (tpm)

expression_summary <- expression %>%
  group_by(sample) %>%
  summarize(total_rpk = sum(rpk, na.rm = T))
expression_summary$rpk_m <- expression_summary$total_rpk / 1e6
expression <- merge(expression, expression_summary)
expression$tpm <-expression$rpk / expression$rpk_m
tpm <- expression %>% group_by(gene_id) %>%
  summarize(tpm = mean(tpm, na.rm = T))
```

Calling peaks and counting overlaps with promoters for lncRNAs and mRNAs

```{r}
# Read in consensus peaks as a list.
consensus_peaks <- import_peaks("/Shares/rinn_class/data/CLASS_2020/analysis/01_consensus_peaks/results/")

# Read in gencode.
gtf <- rtracklayer::import("/Shares/rinn_class/data/genomes/human/gencode/v32/gencode.v32.annotation.gtf")

# Get the promoter regions of each lncRNA and mRNA.

proms <- get_promoter_regions(gencode_gr = gtf, biotype = c("lncRNA", "protein_coding"),
                              upstream =3000, downstream = 3000)

# Count peaks over promoters

promoter_peak_count <- count_peaks_per_feature(features = proms, peak_list = consensus_peaks )
```

Finding ghosts
Reservoirs/Ghost promoters: Promoters with TFs binding but without expression (Note: Here in the script we are using the term "ghosts" and they are same as the reservoirs)
```{r}
# Make this binary -- a co-occurence matrix (the promoters bound by TFs).
promoter_peak_occurence <- matrix(as.numeric(promoter_peak_count > 0),
                           nrow = dim(promoter_peak_count)[1],
                           ncol = dim(promoter_peak_count)[2])
rownames(promoter_peak_occurence) <- rownames(promoter_peak_count)
colnames(promoter_peak_occurence) <- colnames(promoter_peak_count)

# Make a data frame of binding events per promoter.
peak_occurence_df <- data.frame("gene_id" = colnames(promoter_peak_occurence),
                                "number_of_tfs" = colSums(promoter_peak_occurence))

# Merge in the expression data into the data frame we just created.
peak_occurence_df <- merge(peak_occurence_df, tpm)

# Defining ghosts here
ghosts <- peak_occurence_df %>% filter(number_of_tfs > 7, tpm < 0.001)
## subset columns that correspond to ghosts
ghost_peak_occurence <- promoter_peak_occurence[, colnames(promoter_peak_occurence) %in% ghosts$gene_id]

# not_ghosts: everything else
not_ghosts <- peak_occurence_df %>% filter(number_of_tfs > 7, tpm > 0.001)
not_ghost_peak_occurence <- promoter_peak_occurence[,colnames(promoter_peak_occurence) %in% not_ghosts$gene_id]
```

Making a contingency table here
```{r}
# No. of ghost promoters bound by each TF
ghost_num_overlaps <- rowSums(ghost_peak_occurence)
not_ghost_num_overlaps <- rowSums(not_ghost_peak_occurence)

num_peaks <- rowSums(promoter_peak_occurence[,colnames(promoter_peak_occurence)%in% c(colnames(ghost_peak_occurence), colnames(not_ghost_peak_occurence))])
total_ghosts <- length(ghosts)
total_not_ghosts <- length(not_ghosts)
total <- total_ghosts + total_not_ghosts

chisq_res <- data.frame("tf" =character(),
                        "ghosted" = numeric(),
                        "chisq_stat" = numeric(),
                        "chisq_pval" = numeric(),
                        "not_ghosts" = numeric())

# Making a conntingency table for each trancription factor (in num_peaks list)
for(i in 1:length(num_peaks)) {
  total_peaks <- num_peaks[[i]]
  df1 <- data.frame("ghosted" = c("ghost", "ghost", "not_ghost", "not_ghost"),
                    "promoter_bound" = c("bound", "not_bound"),
                    "counted" = c(ghost_num_overlaps[[i]],
                                  total_peaks - ghost_num_overlaps[[i]],
                                  not_ghost_num_overlaps[[i]], 
                                  total_peaks - not_ghost_num_overlaps[[i]])) %>%
    pivot_wider(names_from = ghosted, values_from = counted) %>%
    column_to_rownames("promoter_bound") %>%
    as.matrix()
  # Performing Chi-squared test for ghost vs not_ghost
  csres <- chisq.test(df1)
  phi_coef <- phi(df1)
  tdf <- data.frame("tf" = names(num_peaks[i]),
                    "ghosted" = ghost_num_overlaps[[i]],
                    "not_ghosts" = not_ghost_num_overlaps[[i]],
                    "chisq_stat" = csres$statistic,
                    "chisq_pval" = csres$p.value,
                    "phi_coefficient" = phi_coef$phi)
  
  chisq_res <- bind_rows(chisq_res, tdf)
}
```

Revisit this chunk once the manuscript is ready
```{r}
# TODO revisit this chu
# Use what you just made to plot the results.

# heat map for tf and ghosts
# g <- ggplot(chisq_res)
# g + geom_tile(aes(x = ghosted, y = tf, fill = ghosted), color = "black", size = 0.1) +
#   theme(axis.text = element_text(size = 5)) +
#   scale_fill_gradientn(colors = c("white", "red", "orange", "yellow", "green","turquoise", "blue", "purple"))

g <- ggplot(chisq_res, aes(y = ghosted, x = tf))
g + geom_point(alpha = 0.2) +
  geom_smooth() +
  # scale_y_continuous(expand = c(0,0)) +
  # scale_x_continuous(expand = c(0,0)) +
  # scale_color_manual(values = c("#424242","#a8404c"), name = "Gene type") +
  ggtitle("Number of ghosts per TF") +
  xlab(expression('Number of TFs')) +
  ylab(expression(ghosted))
ggsave("figures/k562_promoter_binding_vs_expression.png")
ggsave("figures/k562_promoter_binding_vs_expression.pdf")
```
