---
title: "Reservoir histone mark metaplots"
output: html_document
editor_options:
  chunk_output_type: console
---
```{r setup, include=FALSE}
options(stringsAsFactors = FALSE)
knitr::opts_chunk$set(echo = TRUE)
# install.packages("tidyverse")
library(tidyverse)
library(GenomicRanges)
library(ChIPseeker)
#library(valr)
```
```{r retrieve-encode-peak-files}
# K9me1 - https://www.encodeproject.org/files/ENCFF285EKW/@@download/ENCFF285EKW.bed.gz
# K27me3 - https://www.encodeproject.org/files/ENCFF031FSF/@@download/ENCFF031FSF.bed.gz
# K27ac - https://www.encodeproject.org/files/ENCFF038DDS/@@download/ENCFF038DDS.bed.gz
# K9me3 - https://www.encodeproject.org/files/ENCFF371GMJ/@@download/ENCFF371GMJ.bed.gz
# K4me1 - https://www.encodeproject.org/files/ENCFF159VKJ/@@download/ENCFF159VKJ.bed.gz
```

#This needs to be cleaned up a lot on Monday

```{r}
# Note, you'll need to append /Shares/rinn_class/data/CLASS_2020/analysis" to the files paths to get the files.
promoters <- rtracklayer::import("/Shares/rinn_class/data/CLASS_2020/analysis/01_consensus_peaks/results/lncrna_mrna_promoters.gtf")
reservoir_promoters <- rtracklayer::import("/Shares/rinn_class/data/CLASS_2020/analysis/04_correlation_with_expression/results/k562_ghost_promoters.bed")
peak_files <- data.frame("file" = list.files("/Shares/rinn_class/data/CLASS_2020/analysis", full.names = TRUE, pattern = ".bed"),
                         "histone_mark" = c("H3K27me3", "H3K27ac", "H3K4me1", "H3K9me1", "H3K9me3"))
# Read in peaks
se_hg38 <- rtracklayer::import("/Shares/rinn_class/data/CLASS_2020/analysis/07_reservoir_properties/results/k562_superenhancers_hg38.bed")

hmark_peaks <- lapply(peak_files$file, read.table)
names(hmark_peaks) <- peak_files$histone_mark
hmark_granges <- lapply(hmark_peaks, function(peaks) {
  GRanges(seqnames = peaks$V1,
          ranges = IRanges(start = peaks$V2,
                           end = peaks$V3))
})

#read in nascent transcription data
nascent_tpm <- read.csv("/Shares/rinn_class/data/CLASS_2020/analysis/06_nascent_expression/results/k562_nascent_tpm.csv")

```


```{r}
# create a matrix indicating whether each chromatin mark overlaps each promoter
hmark_occurence <- matrix(0, nrow = length(reservoir_promoters), ncol = nrow(peak_files))
reservoir_genes <- gsub("_promoter_tss6kb", "", reservoir_promoters$name)
rownames(hmark_occurence) <- reservoir_genes
colnames(hmark_occurence) <- peak_files$histone_mark


# Fill in the matrix with overlaps with reservoirs
for(i in 1:length(hmark_granges)){
  ov <- findOverlaps(reservoir_promoters, hmark_granges[[i]])
  res_overlaps <- unique(ov@from)
  hmark_occurence[res_overlaps, i] <- 1
}

#make super enhancers data frame
# create a matrix indicating whether each super enhancer overlaps each promoter

se_occurence <- matrix(0, nrow = length(reservoir_promoters), ncol = 1)
rownames(se_occurence) <- reservoir_genes
colnames(se_occurence) <- "super_enhancers"

se_overlap <- findOverlaps(reservoir_promoters, se_hg38)
res_overlaps <- unique(se_overlap@from)
se_occurence[res_overlaps,1] <- 1
table(se_occurence[,1])

#Merge histone mark occurence and super enhancer occurence
se_occurence <- as.data.frame(se_occurence) %>% rownames_to_column("GeneID")
hmark_occurence<- as.data.frame(hmark_occurence) %>% rownames_to_column("GeneID")
hmark_se_occurence <- merge(hmark_occurence, se_occurence)
hmark_se_occurence_matrix <- hmark_se_occurence %>% column_to_rownames("GeneID") %>% as.matrix()

#plotting histone occurence and super enhancer occurence
library(eulerr)
genes.venn <- euler(hmark_se_occurence_matrix[,c(-3,-4,-5)])
plot(genes.venn, quantities = TRUE)

```


```{r}
k27_hmark_se_occurence <- hmark_se_occurence_matrix [,c(-3,-4,-5)]
k27_hmark_se_occurence <- as.data.frame(k27_hmark_se_occurence)
H3_K27_me3_and_ac <- data.frame("H3_K27_me3_and_ac" = as.integer())
k27_hmark_se_occurence <- k27_hmark_se_occurence %>% rownames_to_column("GeneID")

for (i in 1:nrow(k27_hmark_se_occurence)) {
  if (k27_hmark_se_occurence$H3K27ac[[i]]==1 & k27_hmark_se_occurence$H3K27me3[[i]] == 1 ) {
    temp_df <- data.frame("GeneID" = k27_hmark_se_occurence$GeneID[[i]], "H3_K27_me3_and_ac" = 1)
    H3_K27_me3_and_ac <- bind_rows(temp_df, H3_K27_me3_and_ac)
  }}
k27_sum_df <- data.frame("H3K27me3_reservoirs" = sum (k27_hmark_se_occurence$H3K27me3),
                         "H3K27ac_reservoirs" = sum (k27_hmark_se_occurence$H3K27ac),
                         "me3_ac_overlap" = sum (H3_K27_me3_and_ac$H3_K27_me3_and_ac),
                         "super_enhancers_reservoirs" = sum (k27_hmark_se_occurence$super_enhancers))
rownames(k27_sum_df) <- "Number_reservoirs"

#Save Overlap Table as CSV
write.csv(k27_sum_df, "/scratch/Users/sosh9992/CLASS_2020/analysis/07_reservoirs/results/H3K27me3_H3K27ac_overlap.csv")
```

```{r}
# add in nascent trancription tpm to dataframe with histone marks
colnames(nascent_tpm) <- c("GeneID", "GeneName", "tpm")
hm_se_nascent <- merge(k27_hmark_se_occurence, nascent_tpm)

histov_nascent <- data.frame("H3K27_me3_and_ac" = as.integer())
for (i in 1:nrow(hm_se_nascent)) {
  if (hm_se_nascent$H3K27ac[[i]]==1 & hm_se_nascent$H3K27me3[[i]] == 1 ) {
    temp_df <- data.frame("GeneID" = hm_se_nascent$GeneID[[i]], "H3K27_me3_and_ac" = 1)
  }
    else{
      temp_df <- data.frame("GeneID" = hm_se_nascent$GeneID[[i]], "H3K27_me3_and_ac" = 0)
    }
    histov_nascent <- bind_rows(temp_df, histov_nascent)
  }

hm_se_nascent <- merge(hm_se_nascent, histov_nascent)

nascent_ghost <- filter(hm_se_nascent, tpm < 0.001)
nascent_zombie <- filter(hm_se_nascent, tpm > 0.001)

nas_me3_ghost <- filter(nascent_ghost, H3K27me3 == 1)
nas_ac_ghost <- filter(nascent_ghost, H3K27ac == 1)
nas_me3_ac_ghost <- filter(nascent_ghost, H3K27_me3_and_ac == 1)

nas_me3_zombie <- filter(nascent_zombie, H3K27me3 == 1)
nas_ac_zombie <- filter(nascent_zombie, H3K27ac == 1)
nas_me3_ac_zombie <- filter(nascent_zombie, H3K27_me3_and_ac == 1)


#Make a density plot for nascent transctiption and histone marks for ghosts
## H3K27me3
g <- ggplot(nas_me3_zombie, aes(x = tpm))
g + geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
  xlab("Nascent_tpm") +
    theme(legend.title = element_text(size = 5), legend.text = element_text(size = 5), legend.key.size = unit(0.5, "cm"), plot.title = element_text(size = 10), axis.title = element_text(size = 7), axis.text.x = element_text(size = 5), axis.text.y = element_text(size = 5)) +
  ggtitle("H3K27me3_nascent")

## H3K27ac
g <- ggplot(nas_ac_zombie, aes(x = log10(tpm)))
g + geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
  xlab("log10_Nascent_tpm") +
    theme(legend.title = element_text(size = 5), legend.text = element_text(size = 5), legend.key.size = unit(0.5, "cm"), plot.title = element_text(size = 10), axis.title = element_text(size = 7), axis.text.x = element_text(size = 5), axis.text.y = element_text(size = 5)) +
  ggtitle("H3K27ac_nascent")

## both acetyl and trimethyl
g <- ggplot(nas_me3_ac_zombie, aes(x = tpm))
g + geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
  xlab("Nascent_tpm") +
    theme(legend.title = element_text(size = 5), legend.text = element_text(size = 5), legend.key.size = unit(0.5, "cm"), plot.title = element_text(size = 10), axis.title = element_text(size = 7), axis.text.x = element_text(size = 5), axis.text.y = element_text(size = 5)) +
  ggtitle("H3K27ac_H3K27me3_nascent")

```


```{r}
# We are going to start out with a function to import files called "import_peaks"
# This the filtered peak_list; final list of 161 TFs after filtering
peak_list <- import_peaks("/Shares/rinn_class/data/CLASS_2020/analysis/01_consensus_peaks/results/consensus_peaks")

# Making the features (gencode) GRanges that has lots of features annotated
gencode_gr <- rtracklayer::import("/Shares/rinn_class/data/genomes/human/gencode/v32/gencode.v32.annotation.gtf")

# Here we are using the function "get_promoter_regions" to grab both the lncRNA and mRNA promoters
proms <- get_promoter_regions(gencode_gr, biotype = c("lncRNA", "protein_coding"))

# Counts the number of promoters that each TF binds to
promoter_peak_count <- count_peaks_per_feature(proms, peak_list)

# Make this binary -- a co-occurence matrix (the promoters bound by TFs).
promoter_peak_occurence <- matrix(as.numeric(promoter_peak_count > 0),
                           nrow = dim(promoter_peak_count)[1],
                           ncol = dim(promoter_peak_count)[2])
rownames(promoter_peak_occurence) <- rownames(promoter_peak_count)
colnames(promoter_peak_occurence) <- colnames(promoter_peak_count)

# Make a data frame of binding events per promoter.
peak_occurence_df <- data.frame("GeneID" = colnames(promoter_peak_occurence),
                                "number_of_tfs" = colSums(promoter_peak_occurence))

# Merge peak occurence into histone mark, SE, nascent data frame
dbp_hm_nascent <- merge(hm_se_nascent, peak_occurence_df)
```


```{r}
#graph number of binding events with or without nascent transcription in density plot
nascent <- data.frame("nascent_exp" = as.character())
for (i in 1:nrow(dbp_hm_nascent)) {
  if (dbp_hm_nascent$tpm[[i]] > 0.001 ) {
    temp_df <- data.frame("GeneID" = dbp_hm_nascent$GeneID[[i]], "nascent_exp" = "yes")
  }
    else{
      temp_df <- data.frame("GeneID" = dbp_hm_nascent$GeneID[[i]], "nascent_exp" = "no")
    }
    nascent <- bind_rows(temp_df, nascent)
  }
#merge nascent with dbps
dbp_nascent <- merge(nascent, peak_occurence_df)
#plot it
g <- ggplot(dbp_nascent, aes(x = number_of_tfs, color = nascent_exp))
g + geom_density() +
  scale_color_manual(values = c("#424242","#a8404c"), name = "nascent_exp")

#same thing for H3K27me3
k27me3 <- data.frame("H3K27me3" = as.character())
for (i in 1:nrow(dbp_hm_nascent)) {
  if (dbp_hm_nascent$H3K27me3[[i]] == 1 ) {
    temp_df <- data.frame("GeneID" = dbp_hm_nascent$GeneID[[i]], "H3K27me3" = "yes")
  }
    else{
      temp_df <- data.frame("GeneID" = dbp_hm_nascent$GeneID[[i]], "H3K27me3" = "no")
    }
    k27me3 <- bind_rows(temp_df, k27me3)
}
#merge in tpm
tpm_k27me3 <- merge(k27me3, nascent_tpm)
#plot it
g <- ggplot(tpm_k27me3, aes(x = log10(tpm), color = H3K27me3))
g + geom_density() +
  scale_color_manual(values = c("#424242","#a8404c"), name = "H3K27me3")

#same thing for H3K27ac
k27ac <- data.frame("H3K27ac" = as.character())
for (i in 1:nrow(dbp_hm_nascent)) {
  if (dbp_hm_nascent$H3K27ac[[i]] == 1 ) {
    temp_df <- data.frame("GeneID" = dbp_hm_nascent$GeneID[[i]], "H3K27ac" = "yes")
  }
    else{
      temp_df <- data.frame("GeneID" = dbp_hm_nascent$GeneID[[i]], "H3K27ac" = "no")
    }
    k27ac <- bind_rows(temp_df, k27ac)
}
#merge in tpm
tpm_k27ac <- merge(k27ac, nascent_tpm)
#plot it
g <- ggplot(tpm_k27ac, aes(x = log10(tpm), color = H3K27ac))
g + geom_density() +
  scale_color_manual(values = c("#424242","#a8404c"), name = "H3K27ac")


```

```{r}
#PolII
polr2a <- peak_list$POLR2A
polr2b <- peak_list$POLR2B

pol2_occurence <- matrix(0, nrow = length(reservoir_promoters), ncol = 1)
polr2a_ov <- findOverlaps(reservoir_promoters, polr2a)
polr2a_res_overlaps <- unique(ov@from)
pol2_occurence[polr2a_res_overlaps,] <- 1

# polr2b_ov <- findOverlaps(reservoir_promoters, polr2b)
# polr2b_res_overlaps <- unique(ov@from)
# pol2_occurence[polr2b_res_overlaps,] <- 1

rownames(pol2_occurence) <- reservoir_genes
colnames(pol2_occurence) <- c("Polr2a")

# Make a data frame of binding events per promoter.
pol2_occurence_df <- data.frame("GeneID" = rownames(pol2_occurence),
                                "number_bound" = rowSums(pol2_occurence))
#make yes or no
pol2 <- data.frame("Pol2" = as.character())
for (i in 1:nrow(pol2_occurence_df)) {
  if (pol2_occurence_df$number_bound[[i]] > 0) {
    temp_df <- data.frame("GeneID" = pol2_occurence_df$GeneID[[i]], "Pol2" = "yes")
  }
    else{
      temp_df <- data.frame("GeneID" = pol2_occurence_df$GeneID[[i]], "Pol2" = "no")
    }
    pol2 <- bind_rows(temp_df, pol2)
}
#merge in tpm
tpm_pol2 <- merge(pol2, nascent_tpm)

#plot it
g <- ggplot(tpm_pol2, aes(x = log10(tpm), color = Pol2))
g + geom_density() +
  scale_color_manual(values = c("#424242","#a8404c"), name = "Pol2")

```

