---
title: "Reservoir histone mark metaplots"
output: html_document
editor_options:
  chunk_output_type: console
---
```{r setup, include=FALSE}
options(stringsAsFactors = FALSE)
knitr::opts_chunk$set(echo = TRUE)
# install.packages("tidyverse")
library(tidyverse)
library(GenomicRanges)
library(ChIPseeker)
#library(valr)
```
```{r retrieve-encode-peak-files}
# K9me1 - https://www.encodeproject.org/files/ENCFF285EKW/@@download/ENCFF285EKW.bed.gz
# K27me3 - https://www.encodeproject.org/files/ENCFF031FSF/@@download/ENCFF031FSF.bed.gz
# K27ac - https://www.encodeproject.org/files/ENCFF038DDS/@@download/ENCFF038DDS.bed.gz
# K9me3 - https://www.encodeproject.org/files/ENCFF371GMJ/@@download/ENCFF371GMJ.bed.gz
# K4me1 - https://www.encodeproject.org/files/ENCFF159VKJ/@@download/ENCFF159VKJ.bed.gz
```
```{r}
# Note, you'll need to append /Shares/rinn_class/data/CLASS_2020/analysis" to the files paths to get the files.
promoters <- rtracklayer::import("/Shares/rinn_class/data/CLASS_2020/analysis/01_consensus_peaks/results/lncrna_mrna_promoters.gtf")
reservoir_promoters <- rtracklayer::import("/Shares/rinn_class/data/CLASS_2020/analysis/05_rnaseq_expression/results/k562_ghost_promoters.bed")
peak_files <- data.frame("file" = list.files("/scratch/Users/sosh9992/CLASS_2020/analysis/07_reservoirs/", 
                          full.names = TRUE, pattern = ".bed"),
                         "histone_mark" = c("H3K27me3", "H3K27ac", "H3K4me1", "H3K9me1", "H3K9me3"))
# Read in peaks
se_hg38 <- rtracklayer::import("/Shares/rinn_class/data/CLASS_2020/analysis/07_reservoirs/results/k562_superenhancers_hg38.bed")
hmark_peaks <- lapply(peak_files$file, read.table)
names(hmark_peaks) <- peak_files$histone_mark
hmark_granges <- lapply(hmark_peaks, function(peaks) {
  GRanges(seqnames = peaks$V1,
          ranges = IRanges(start = peaks$V2,
                           end = peaks$V3))
})

#read in nascent transcription data
nascent_tpm <- read.csv("/Shares/rinn_class/data/CLASS_2020/analysis/08_nascent/results/k562_nascent_tpm.csv")

```

```{r}
# create a matrix indicating whether each chromatin mark overlaps each promoter
hmark_occurence <- matrix(0, nrow = length(reservoir_promoters), ncol = nrow(peak_files))
reservoir_genes <- gsub("_promoter_tss6kb", "", reservoir_promoters$name)
rownames(hmark_occurence) <- reservoir_genes
colnames(hmark_occurence) <- peak_files$histone_mark


# Fill in the matrix with overlaps with reservoirs
for(i in 1:length(hmark_granges)){
  ov <- findOverlaps(reservoir_promoters, hmark_granges[[i]])
  res_overlaps <- unique(ov@from)
  hmark_occurence[res_overlaps, i] <- 1
}

#make super enhancers data frame
# create a matrix indicating whether each super enhancer overlaps each promoter

se_occurence <- matrix(0, nrow = length(reservoir_promoters), ncol = 1)
rownames(se_occurence) <- reservoir_genes
colnames(se_occurence) <- "super_enhancers"

se_overlap <- findOverlaps(reservoir_promoters, se_hg38)
res_overlaps <- unique(se_overlap@from)
se_occurence[res_overlaps,1] <- 1
table(se_occurence[,1])

#Merge histone mark occurence and super enhancer occurence
se_occurence <- as.data.frame(se_occurence) %>% rownames_to_column("GeneID")
hmark_occurence<- as.data.frame(hmark_occurence) %>% rownames_to_column("GeneID")
hmark_se_occurence <- merge(hmark_occurence, se_occurence)
hmark_se_occurence_matrix <- hmark_se_occurence %>% column_to_rownames("GeneID") %>% as.matrix()

#plotting histone occurence and super enhancer occurence
library(eulerr)
genes.venn <- euler(hmark_se_occurence_matrix[,c(-3,-4,-5)])
plot(genes.venn, quantities = TRUE)

```


```{r}
k27_hmark_se_occurence <- hmark_se_occurence_matrix [,c(-3,-4,-5)]
k27_hmark_se_occurence <- as.data.frame(k27_hmark_se_occurence)
H3_K27_me3_and_ac <- data.frame("H3_K27_me3_and_ac" = as.integer())
k27_hmark_se_occurence <- k27_hmark_se_occurence %>% rownames_to_column("GeneID")

for (i in 1:nrow(k27_hmark_se_occurence)) {
  if (k27_hmark_se_occurence$H3K27ac[[i]]==1 & k27_hmark_se_occurence$H3K27me3[[i]] == 1 ) {
    temp_df <- data.frame("GeneID" = k27_hmark_se_occurence$GeneID[[i]], "H3_K27_me3_and_ac" = 1)
    H3_K27_me3_and_ac <- bind_rows(temp_df, H3_K27_me3_and_ac)
  }}
k27_sum_df <- data.frame("H3K27me3_reservoirs" = sum (k27_hmark_se_occurence$H3K27me3),
                         "H3K27ac_reservoirs" = sum (k27_hmark_se_occurence$H3K27ac),
                         "me3_ac_overlap" = sum (H3_K27_me3_and_ac$H3_K27_me3_and_ac),
                         "super_enhancers_reservoirs" = sum (k27_hmark_se_occurence$super_enhancers))
rownames(k27_sum_df) <- "Number_reservoirs"

#Save Overlap Table as CSV
write.csv(k27_sum_df, "/scratch/Users/sosh9992/CLASS_2020/analysis/07_reservoirs/results/H3K27me3_H3K27ac_overlap.csv")
```

```{r}
# add in nascent trancription tpm to dataframe with histone marks
colnames(nascent_tpm) <- c("GeneID", "GeneName", "tpm")
hm_se_nascent <- merge(k27_hmark_se_occurence, nascent_tpm)

histov_nascent <- data.frame("H3K27_me3_and_ac" = as.integer())
for (i in 1:nrow(hm_se_nascent)) {
  if (hm_se_nascent$H3K27ac[[i]]==1 & hm_se_nascent$H3K27me3[[i]] == 1 ) {
    temp_df <- data.frame("GeneID" = hm_se_nascent$GeneID[[i]], "H3K27_me3_and_ac" = 1)
  }
    else{
      temp_df <- data.frame("GeneID" = hm_se_nascent$GeneID[[i]], "H3K27_me3_and_ac" = 0)
    }
    histov_nascent <- bind_rows(temp_df, histov_nascent)
  }

hm_se_nascent <- merge(hm_se_nascent, histov_nascent)

nascent_ghost <- filter(hm_se_nascent, tpm < 0.001)
nascent_zombie <- filter(hm_se_nascent, tpm > 0.001)

nas_me3_ghost <- filter(nascent_ghost, H3K27me3 == 1)
nas_ac_ghost <- filter(nascent_ghost, H3K27ac == 1)
nas_me3_ac_ghost <- filter(nascent_ghost, H3K27_me3_and_ac == 1)

nas_me3_zombie <- filter(nascent_zombie, H3K27me3 == 1)
nas_ac_zombie <- filter(nascent_zombie, H3K27ac == 1)
nas_me3_ac_zombie <- filter(nascent_zombie, H3K27_me3_and_ac == 1)


#Make a density plot for nascent transctiption and histone marks for ghosts
## H3K27me3
g <- ggplot(nas_me3_zombie, aes(x = tpm))
g + geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
  xlab("Nascent_tpm") +
    theme(legend.title = element_text(size = 5), legend.text = element_text(size = 5), legend.key.size = unit(0.5, "cm"), plot.title = element_text(size = 10), axis.title = element_text(size = 7), axis.text.x = element_text(size = 5), axis.text.y = element_text(size = 5)) +
  ggtitle("H3K27me3_nascent")

## H3K27ac
g <- ggplot(nas_ac_zombie, aes(x = log10(tpm)))
g + geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
  xlab("log10_Nascent_tpm") +
    theme(legend.title = element_text(size = 5), legend.text = element_text(size = 5), legend.key.size = unit(0.5, "cm"), plot.title = element_text(size = 10), axis.title = element_text(size = 7), axis.text.x = element_text(size = 5), axis.text.y = element_text(size = 5)) +
  ggtitle("H3K27ac_nascent")

## both acetyl and trimethyl
g <- ggplot(nas_me3_ac_zombie, aes(x = tpm))
g + geom_density(alpha = 0.2, color = "#424242", fill = "#424242") +
  xlab("Nascent_tpm") +
    theme(legend.title = element_text(size = 5), legend.text = element_text(size = 5), legend.key.size = unit(0.5, "cm"), plot.title = element_text(size = 10), axis.title = element_text(size = 7), axis.text.x = element_text(size = 5), axis.text.y = element_text(size = 5)) +
  ggtitle("H3K27ac_H3K27me3_nascent")

```


---------------
