---
title: "11_reservoir_chromatin_properties"
author: "JR"
date: "5/29/2020"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)

```

#Goal: to determine the chromatin environment of reservoir promoters using histone modification ChIP data. Specifically, Histone H3 (H3) Lysine 27 tri-methylation (H3K27me3) that represents heterochraomtic regions / and H3 lysine 27 acetylation that represents euchromatin.

```{r} 

# Read our compiled peak_occurence data frame from 10_reservoir_nascent_txn (with all the variables for each promoter observation). 
peak_occurrence_df <- read_csv("../10_reservoir_nascent_txn/results/peak_occurrence_df_nascent_added.csv")

# We alos need the peak occurence of promoters and DBPs bound as a matrix.
peak_occurrence_matrix <- read.table("../01_consensus_peaks/results/lncrna_mrna_promoter_peak_occurence_matrix.tsv")


# Histone mark overlaps
# K27me3 - https://www.encodeproject.org/files/ENCFF031FSF/@@download/ENCFF031FSF.bed.gz
# K27ac - https://www.encodeproject.org/files/ENCFF038DDS/@@download/ENCFF038DDS.bed.gz


# Retrieve Histone mark and # TODO add unzip

system("cd data;
           wget https://www.encodeproject.org/files/ENCFF031FSF/@@download/ENCFF031FSF.bed.gz",
           intern = FALSE, ignore.stdout = TRUE, ignore.stderr = TRUE)
system("cd data;
           wget https://www.encodeproject.org/files/ENCFF038DDS/@@download/ENCFF038DDS.bed.gz",
           intern = FALSE, ignore.stdout = TRUE, ignore.stderr = TRUE)



peak_files <- data.frame("file" = c("data/ENCFF031FSF.bed", "data/ENCFF038DDS.bed"),
                         "histone_mark" = c("H3K27me3", "H3K27ac"))
hmark_peaks <- lapply(peak_files$file, read.table)
names(hmark_peaks) <- peak_files$histone_mark
hmark_granges <- lapply(hmark_peaks, function(peaks) {
  GRanges(seqnames = peaks$V1,
          ranges = IRanges(start = peaks$V2,
                           end = peaks$V3))
})



plot(peak_occurrence_df$nascentTpm ~peak_occurrence_df$tpm)
```





```{r}
# create a matrix indicating whether each chromatin mark overlaps each promoter
hmark_occurence <- matrix(0, nrow = length(peak_occurrence_df$reservoir), ncol = nrow(peak_files))
rownames(hmark_occurence) <- peak_occurrence_df$reservoir
colnames(hmark_occurence) <- peak_files$histone_mark
# Fill in the matrix with overlaps
# First column 1
ov1 <- findOverlaps(reservoir_promoters, hmark_granges[[1]])
res_overlaps <- unique(ov1@from)
hmark_occurence[res_overlaps,1] <- 1
table(hmark_occurence[,1])
# Repeat for the next histone marks
ov2 <- findOverlaps(reservoir_promoters, hmark_granges[[2]])
res_overlaps <- unique(ov2@from)
hmark_occurence[res_overlaps,2] <- 1
table(hmark_occurence[,2])
# Repeat for the next histone marks
ov3 <- findOverlaps(reservoir_promoters, hmark_granges[[3]])
res_overlaps <- unique(ov3@from)
hmark_occurence[res_overlaps,3] <- 1
table(hmark_occurence[,3])
# Repeat for the next histone marks
ov4 <- findOverlaps(reservoir_promoters, hmark_granges[[4]])
res_overlaps <- unique(ov4@from)
hmark_occurence[res_overlaps,4] <- 1
table(hmark_occurence[,4])
# Repeat for the next histone marks
ov5 <- findOverlaps(reservoir_promoters, hmark_granges[[5]])
res_overlaps <- unique(ov5@from)
hmark_occurence[res_overlaps,5] <- 1
table(hmark_occurence[,5])
#make super enhancers data frame
# create a matrix indicating whether each super enhancer overlaps each promoter
super_enhacer_res_ov <- rtracklayer::import("/Shares/rinn_class/data/CLASS_2020/analysis/07_reservoirs/results/reservoirs_overlapping_ses.bed")
se_occurence <- matrix(0, nrow = length(reservoir_promoters), ncol = 1)
rownames(se_occurence) <- reservoir_promoters$name
colnames(se_occurence) <- "super_enhancers"
ov6 <- findOverlaps(reservoir_promoters, se_hg19)
res_overlaps <- unique(ov6@from)
se_occurence[res_overlaps,1] <- 1
table(se_occurence[,1])
#Merge histone mark occurence and super enhancer occurence
se_occurence_df <- as.data.frame(se_occurence) %>% rownames_to_column("GeneID")
hmark_occurence_df <- as.data.frame(hmark_occurence) %>% rownames_to_column("GeneID")
hmark_se_occurence_ <- merge(hmark_occurence_df, se_occurence_df)
hmark_se_occurence_matrix <- hmark_se_occurence_ %>% column_to_rownames("GeneID") %>% as.matrix()
#plotting histone occurence and super enhancer occurence
library(eulerr)
genes.venn <- euler(hmark_se_occurence_matrix[,c(-3,-4,-5)])
plot(genes.venn, quantities = TRUE)
```
```{r}
k27_hmark_se_occurence <- hmark_se_occurence_matrix [,c(-3,-4,-5)]
k27_hmark_se_occurence <- as.data.frame(k27_hmark_se_occurence)
H3_K27_me3_and_ac <- data.frame("H3_K27_me3_and_ac" = as.integer())
k27_hmark_se_occurence <- k27_hmark_se_occurence %>% rownames_to_column("GeneID")
for (i in 1:nrow(k27_hmark_se_occurence)) {
  if (k27_hmark_se_occurence$H3K27ac[[i]]==1 & k27_hmark_se_occurence$H3K27me3[[i]] == 1 ) {
    temp_df <- data.frame("GeneID" = k27_hmark_se_occurence$GeneID[[i]], "H3_K27_me3_and_ac" = 1)
    H3_K27_me3_and_ac <- bind_rows(temp_df, H3_K27_me3_and_ac)
  }}
k27_sum_df <- data.frame("H3K27me3_reservoirs" = sum (k27_hmark_se_occurence$H3K27me3),
                         "H3K27ac_reservoirs" = sum (k27_hmark_se_occurence$H3K27ac),
                         "me3_ac_overlap" = sum (H3_K27_me3_and_ac$H3_K27_me3_and_ac),
                         "super_enhancers_reservoirs" = sum (k27_hmark_se_occurence$super_enhancers))
rownames(k27_sum_df) <- "Number_reservoirs"
#Save Overlap Table as CSV
write.csv(k27_sum_df, "H3K27me3_H3K27ac_overlap.csv")
```
---------------
This is some messy draft code. It's not done.
```{r}
# Let's get the reservoir promoter regions
reservoir_promoters <- rtracklayer::import("../04_correlation_with_expression/results/k562_ghost_promoters.bed")
peak_files <- data.frame("file" = list.files("data/", full.names = TRUE, pattern = ".bed"),
                         "histone_mark" = c("H3K27me3", "H3K27ac", "H3K4me1", "H3K9me1", "H3K9me3"))
hmark_peaks <- lapply(peak_files$file, read.table)
names(hmark_peaks) <- peak_files$histone_mark
hmark_granges <- lapply(hmark_peaks, function(peaks) {
  GRanges(seqnames = peaks$V1,
          ranges = IRanges(start = peaks$V2,
                           end = peaks$V3))
})
res_hmarks <- lapply(hmark_granges, function(peaks) {
  subsetByOverlaps(peaks, reservoir_promoters)
})
hmark_peaks <- lapply(peak_files$file, read_bed)
names(hmark_peaks) <- peak_files$histone_mark
res_prom <- read_bed("../04_correlation_with_expression/results/k562_ghost_promoters.bed")
res_prom_windows <- res_prom %>%
  bed_makewindows(50)
res_overlaps <- bed_map(res_prom_windows, hmark_peaks[[2]], .counts = dplyr::n())
res_overlaps <- res_overlaps %>%
  group_by(.win_id) %>%
  summarize(sum_counts = sum(.counts, na.rm = T))
g <- ggplot(res_overlaps, aes(x = .win_id, y = sum_counts))
g + geom_point()
names(hmark_peaks)
me3_ac_overlaps <- bed_intersect(hmark_peaks[[1]], hmark_peaks[[2]])
nrow(hmark_peaks[[1]])
nrow(hmark_peaks[[2]])
res_me3ac_overlaps <- bed_intersect(res_prom, hmark_peaks[[1]])
res_me3ac_overlaps <- bed_intersect(res_prom, hmark_peaks[[2]])