---
title: "11_reservoir_chromatin_properties"
author: "JR"
date: "5/29/2020"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
source("../util/intersect_functions.R")
source("../util/_setup.R")

```

#Goal: to determine the chromatin environment of reservoir promoters using histone modification ChIP data. Specifically, Histone H3 (H3) Lysine 27 tri-methylation (H3K27me3) that represents heterochraomtic regions / and H3 lysine 27 acetylation that represents euchromatin.

```{r} 

# First we will go get the .bed files of significant peaks determined by encode in k562
# K27me3 - https://www.encodeproject.org/files/ENCFF031FSF/@@download/ENCFF031FSF.bed.gz
# K27ac - https://www.encodeproject.org/files/ENCFF038DDS/@@download/ENCFF038DDS.bed.gz


# Retrieve Histone mark and # TODO add unzip

system("cd data;
           wget https://www.encodeproject.org/files/ENCFF031FSF/@@download/ENCFF031FSF.bed.gz",
           intern = FALSE, ignore.stdout = TRUE, ignore.stderr = TRUE)
system("cd data;
           wget https://www.encodeproject.org/files/ENCFF038DDS/@@download/ENCFF038DDS.bed.gz",
           intern = FALSE, ignore.stdout = TRUE, ignore.stderr = TRUE)

system("cd data;
       gunzip *.gz")


# Organzing the histone mark data into a data frame and then granges of start and stop of peaks in .bed files

peak_files <- data.frame("file" = c("data/ENCFF031FSF.bed", "data/ENCFF038DDS.bed"),
                         "histone_mark" = c("H3K27me3", "H3K27ac"))
hmark_peaks <- lapply(peak_files$file, read.table)
names(hmark_peaks) <- peak_files$histone_mark
hmark_granges <- lapply(hmark_peaks, function(peaks) {
  GRanges(seqnames = peaks$V1,
          ranges = IRanges(start = peaks$V2,
                           end = peaks$V3))
         })

# To find overlaps with promoters using gRanges we will open up the GTF of promoters from 01_consensus_peaks
promoters <- rtracklayer::import("../01_consensus_peaks/results/lncrna_mrna_promoters.gtf")

# Read our compiled peak_occurrence data frame from 10_reservoir_nascent_txn (with all the variables for each promoter observation). 
peak_occurrence_df <- read_csv("../10_reservoir_nascent_txn/results/peak_occurrence_df_nascent_added.csv")

# create a matrix indicating whether each chromatin mark overlaps each promoter
hmark_occurrence <- matrix(0, nrow = length(peak_occurrence_df$gene_id), ncol = nrow(peak_files))
rownames(hmark_occurrence) <- peak_occurrence_df$gene_id
colnames(hmark_occurrence) <- peak_files$histone_mark



# K27me3 Promoter overlaps
ov1 <- findOverlaps(promoters, hmark_granges[[1]])
overlaps <- unique(ov1@from)
hmark_occurrence[overlaps,1] <- 1
table(hmark_occurrence[,1])

# K27ac Promoter overlaps
ov2 <- findOverlaps(promoters, hmark_granges[[2]])
overlaps <- unique(ov2@from)
hmark_occurrence[overlaps,2] <- 1
table(hmark_occurrence[,2])

# Making a data frame of all promoters K27me3 and K27ac status as binary integers.
# Changing row names to be compatible with peak_occurrence_df
hmark_occurrence_df <- as.data.frame(hmark_occurrence) %>% rownames_to_column("gene_id")

# merging into peak_occurence_df histone calls by gene_id
peak_occurrence_df_histone <- merge(peak_occurrence_df, hmark_occurrence_df, by = "gene_id")

# Writting out final peak_occurrence_df !

write.csv(peak_occurrence_df_histone, "results/peak_occurence_df.csv")

```

#TODO I spot checked and couldn't find the first test example :

	
ENSG00000002822.15

I looked up coordinates in UCSC and then tried to find a window in bed file -- didn't seem to overlap?




#START HERE

```{r}
#plotting histone occurrence and super enhancer occurrence
library(eulerr)
genes.venn <- euler(hmark_se_occurrence_matrix[,c(-3,-4,-5)])
plot(genes.venn, quantities = TRUE)




```
---------------
