---
title: "11_reservoir_chromatin_properties"
author: "JR"
date: "5/29/2020"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)

```

#Goal: to determine the chromatin environment of reservoir promoters using histone modification ChIP data. Specifically, Histone H3 (H3) Lysine 27 tri-methylation (H3K27me3) that represents heterochraomtic regions / and H3 lysine 27 acetylation that represents euchromatin.

```{r} 

# Read our compiled peak_occurence data frame from 10_reservoir_nascent_txn (with all the variables for each promoter observation). 
peak_occurrence_df <- read_csv("../10_reservoir_nascent_txn/results/peak_occurrence_df_nascent_added.csv")

# We alos need the peak occurence of promoters and DBPs bound as a matrix.
peak_occurrence_matrix <- read.table("../01_consensus_peaks/results/lncrna_mrna_promoter_peak_occurence_matrix.tsv")


# Histone mark overlaps
# K27me3 - https://www.encodeproject.org/files/ENCFF031FSF/@@download/ENCFF031FSF.bed.gz
# K27ac - https://www.encodeproject.org/files/ENCFF038DDS/@@download/ENCFF038DDS.bed.gz


# Retrieve Histone mark and # TODO add unzip

system("cd data;
           wget https://www.encodeproject.org/files/ENCFF031FSF/@@download/ENCFF031FSF.bed.gz",
           intern = FALSE, ignore.stdout = TRUE, ignore.stderr = TRUE)
system("cd data;
           wget https://www.encodeproject.org/files/ENCFF038DDS/@@download/ENCFF038DDS.bed.gz",
           intern = FALSE, ignore.stdout = TRUE, ignore.stderr = TRUE)

system("cd data;
       gunzip *.gz")


# Organzing the histone mark data into a data frame and then granges of start and stop of peaks in .bed files

peak_files <- data.frame("file" = c("data/ENCFF031FSF.bed", "data/ENCFF038DDS.bed"),
                         "histone_mark" = c("H3K27me3", "H3K27ac"))
hmark_peaks <- lapply(peak_files$file, read.table)
names(hmark_peaks) <- peak_files$histone_mark
hmark_granges <- lapply(hmark_peaks, function(peaks) {
  GRanges(seqnames = peaks$V1,
          ranges = IRanges(start = peaks$V2,
                           end = peaks$V3))

          })



# create a matrix indicating whether each chromatin mark overlaps each promoter
hmark_occurence <- matrix(0, nrow = length(peak_occurrence_df$gene_id), ncol = nrow(peak_files))
rownames(hmark_occurence) <- peak_occurrence_df$gene_id
colnames(hmark_occurence) <- peak_files$histone_mark



# K27me3 Promoter overlaps
ov1 <- findOverlaps(promoters, hmark_granges[[1]])
overlaps <- unique(ov1@from)
hmark_occurence[overlaps,1] <- 1
table(hmark_occurence[,1])

# K27ac Promoter overlaps
ov2 <- findOverlaps(promoters, hmark_granges[[2]])
overlaps <- unique(ov2@from)
hmark_occurence[overlaps,2] <- 1
table(hmark_occurence[,2])

# Making a data frame of all promoters K27me3 and K27ac status as binary integers.

hmark_occurence_df <- as.data.frame(hmark_occurence) %>% rownames_to_column("GeneID")



```






#START HERE

```{r}
#plotting histone occurence and super enhancer occurence
library(eulerr)
genes.venn <- euler(hmark_se_occurence_matrix[,c(-3,-4,-5)])
plot(genes.venn, quantities = TRUE)




```
---------------
