---
title: "TE Properties"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
library(GenomicFeatures)
library(rtracklayer)
library(ggrepel)
library(ggpubr)
library(regioneR)
library(Rsubread)
source("../util/intersect_functions.R")
source("../util/_setup.R")
```

```{r import}
base_path <- "../02_permutation_of_consensus_peaks/results"
perm_res_family <- read.csv(file.path(base_path,
                                      "permutation_results_repeat_families.csv"))
perm_res_class <- read.csv(file.path(base_path,
                                     "permutation_results_repeat_classes.csv"))
```

```{r classes-clustering, fig.height=4, fig.width=25, message=FALSE}
class_ov_matrix <- perm_res_class %>%
  dplyr::select(region, tf, zscore) %>%
  pivot_wider(names_from = tf, values_from = zscore) %>%
  column_to_rownames("region") %>%
  as.matrix()
class_region_clust <- hclust(dist(class_ov_matrix))
class_tf_clust <- hclust(dist(t(class_ov_matrix)))
plot(class_region_clust)

perm_res_class$region <- factor(perm_res_class$region,  
                                class_region_clust$labels[class_region_clust$order])
perm_res_class$tf <- factor(perm_res_class$tf,  
                           class_tf_clust$labels[class_tf_clust$order])


#### FIGURE: Figure 3A
g <- ggplot(perm_res_class, aes(x = tf, y = region, fill = zscore))
g + geom_tile() + 
  scale_fill_gradient2(limits = c(-100, 100), na.value = "#ffffff") +  
  theme(axis.text.x = element_text(angle = 90L, hjust = 1L, vjust = 0.5))
ggsave("figures/dbp_binding_on_repeat_classes.png", height = 4, width = 25)
ggsave("figures/dbp_binding_on_repeat_classes.pdf", height = 4, width = 25)
```

```{r families-clustering, fig.height=9, fig.width=25, message=FALSE}
# Let's filter out regions that have no significant enrichments or depletions
sig_overlaps <- perm_res_family %>% filter(padj < 0.01)
sig_regions <- unique(as.character(sig_overlaps$region))

family_ov_matrix <- perm_res_family %>%
  filter(!is.na(zscore), !is.na(region),
         region %in% sig_regions) %>%
  dplyr::select(region, tf, zscore) %>%
  pivot_wider(names_from = tf, values_from = zscore) %>%
  column_to_rownames("region") %>%
  as.matrix()
family_region_clust <- hclust(dist(family_ov_matrix))
family_tf_clust <- hclust(dist(t(family_ov_matrix)))
plot(family_region_clust)

subset_family <- perm_res_family %>% filter(!is.na(zscore), !is.na(region),
         region %in% sig_regions)

subset_family$region <- factor(subset_family$region,  
                                family_region_clust$labels[family_region_clust$order])
subset_family$tf <- factor(subset_family$tf,  
                           family_tf_clust$labels[family_tf_clust$order])


#### FIGURE: Figure 3A
g <- ggplot(subset_family, aes(x = tf, y = region, fill = zscore))
g + geom_tile() + 
  scale_fill_gradient2(limits = c(-100, 100), na.value = "#ffffff") +  
  theme(axis.text.x = element_text(angle = 90L, hjust = 1L, vjust = 0.5))
ggsave("figures/dbp_binding_on_repeat_family.png", height = 9, width = 25)
ggsave("figures/dbp_binding_on_repeat_family.pdf", height = 9, width = 25)
```

# SVA family enriched DBPs

```{r sva-family}
sva_enriched <- subset_family %>% 
  filter(region == "SVA_family",
         padj < 0.01) %>%
  arrange(-zscore)

# Let's take the top three
sva_enriched_dbp <- sva_enriched %>%
  slice(1:3) %>%
  pull("tf") %>%
  as.character()
sva_enriched_dbp

# Let's look at some overlaps with SVA and thes DBPs
peak_list <- import_peaks("../01_consensus_peaks/results/consensus_peaks/filtered_by_peaks/")
poi <- peak_list[sva_enriched_dbp]

rmsk <- import_repeatmasker()
rmsk_sva <- rmsk[rmsk$rep_family == "SVA"]

rmsk_CBFA2T3_ov <- findOverlaps(rmsk_sva,
                                peak_list[[sva_enriched_dbp[[1]]]])
rmsk_sva$CBFA2T3_ov <- FALSE
rmsk_sva$CBFA2T3_ov[unique(rmsk_CBFA2T3_ov@from)] <- TRUE

rmsk_ZBTB33_ov <- findOverlaps(rmsk_sva,
                                peak_list[[sva_enriched_dbp[[2]]]])
rmsk_sva$ZBTB33_ov <- FALSE
rmsk_sva$ZBTB33_ov[unique(rmsk_ZBTB33_ov@from)] <- TRUE


rmsk_CBFA2T2_ov <- findOverlaps(rmsk_sva,
                                peak_list[[sva_enriched_dbp[[3]]]])
rmsk_sva$CBFA2T2_ov <- FALSE
rmsk_sva$CBFA2T2_ov[unique(rmsk_CBFA2T2_ov@from)] <- TRUE


rmsk_sva_all_ov <- rmsk_sva[rmsk_sva$CBFA2T2_ov & 
                              rmsk_sva$ZBTB33_ov & 
                              rmsk_sva$CBFA2T3_ov]
rtracklayer::export(rmsk_sva_all_ov,
                    "results/sva_dbp_top3_overlaps.bed")

promoters <- rtracklayer::import("../01_consensus_peaks/results/lncrna_mrna_promoters.gtf")

sva_promoter_ov <- findOverlaps(rmsk_sva_all_ov, promoters)
sva_dbp_promoter <- rmsk_sva_all_ov[unique(sva_promoter_ov@from)]
rtracklayer::export(sva_dbp_promoter, 
                    "results/sva_dbp_top3_promoter_overlaps.bed")
```

## SVA expression levels in RNA-seq and Nascent-seq

```{r sva-erv-saf}
sva_erv_list <- c("ERVL-MaLR", "ERV1", "ERVK", "ERVL", "SVA")
# Let's make a SAF file for SVAs and ERVs
repeats_sva_erv <- rmsk[rmsk$rep_family %in% sva_erv_list] %>%
  as.data.frame()

# Let's see if the widths are comparable
g <- ggplot(repeats_sva_erv, aes(x = log10(width+1), fill = rep_family))
g + geom_density(alpha = 0.4) + 
  ggtitle("Width of selected repeat elements")

# Okay, now let's actually make the SAF
# We'll make the gene name the repeat name plus the repeat family plus an index
# for uniquely identifying it. 
repeats_sva_erv <- repeats_sva_erv %>%
  group_by(rep_family, rep_name) %>%
  mutate(index = 1:n())

repeats_sva_erv <- repeats_sva_erv %>%
  unite("gene_id", rep_family, rep_name, index,
        sep = ";")
 
# Now let's grab just the columns we need
repeats_sva_erv <- repeats_sva_erv %>%
  dplyr::select("gene_id", "seqnames", "start", "end", "strand") 
names(repeats_sva_erv) <- c("GeneID", "Chr", "Start", "End", "Strand")

write.table(repeats_sva_erv, "results/rmsk_sva_erv.SAF", 
            sep = "\t",
            quote = F, row.names = F)
```

```{r sva-rna-seq}
counts <- featureCounts(c("../07_binding_versus_expression/data/ENCFF625ZBS.bam", 
                          "../07_binding_versus_expression/data/ENCFF630HEX.bam"),
                          annot.ext = "results/rmsk_sva_erv.SAF",
                          isPairedEnd = TRUE,
                          nthreads = 16)

rpk <- counts$counts / (counts$annotation$Length/1000)
expression <- data.frame("rpk" = rpk) %>%
    rownames_to_column("gene_id") %>%
    pivot_longer(2:3, names_to = "sample", values_to = "rpk")
  
expression_summary <- expression %>%
    group_by(sample) %>%
    summarize(total_rpk = sum(rpk, na.rm = T))
expression_summary$rpk_m <- expression_summary$total_rpk / 1e6
expression <- merge(expression, expression_summary)
expression$tpm <- expression$rpk / expression$rpk_m
tpm <- expression %>% group_by(gene_id) %>%
    summarize(rnaseq_tpm = mean(tpm, na.rm = T))
```


```{r sva-nascent-seq}
counts <- featureCounts(c("../10_reservoir_nascent_txn/data/516.sorted.bam", 
                          "../10_reservoir_nascent_txn/data/829.sorted.bam"), 
                          annot.ext = "results/rmsk_sva_erv.SAF", 
                          isPairedEnd = TRUE, nthreads = 16)
rpk <- counts$counts / (counts$annotation$Length/1000)
expression <- data.frame("rpk" = rpk) %>%
    rownames_to_column("gene_id") %>%
    pivot_longer(2:3, names_to = "sample", values_to = "rpk")
  
expression_summary <- expression %>%
    group_by(sample) %>%
    summarize(total_rpk = sum(rpk, na.rm = T))
expression_summary$rpk_m <- expression_summary$total_rpk / 1e6
expression <- merge(expression, expression_summary)
expression$tpm <- expression$rpk / expression$rpk_m
nascent_tpm <- expression %>% group_by(gene_id) %>%
    summarize(nascent_tpm = mean(tpm, na.rm = T))

combined_tpm <- merge(tpm, nascent_tpm)

# Split back out families
combined_tpm <- combined_tpm %>%
  separate(gene_id, into = c("rep_family", "rep_name", "rep_index"), 
           remove = F, sep = ";")

# Make a category for ERV vs. SVA
combined_tpm$fam <- "ERV"
combined_tpm[combined_tpm$rep_family == "SVA", "fam"] <- "SVA"
table(combined_tpm$fam)
```

```{r rep-exp-plots}
g <- ggplot(combined_tpm, aes(x = log10(rnaseq_tpm + 0.1), fill = rep_family))
g + geom_density(alpha = 0.4) + 
  ggtitle("RNA-seq expression by repeat family")

g <- ggplot(combined_tpm, aes(x = rep_family, y = log10(rnaseq_tpm + 0.1)))
g + geom_boxplot() + 
  stat_compare_means() +
  ggtitle("RNA-seq expression by repeat family")

g <- ggplot(combined_tpm, aes(x = fam, y = log10(rnaseq_tpm + 0.1)))
g + geom_boxplot() + 
  stat_compare_means() +
  ggtitle("RNA-seq expression by repeat")


g <- ggplot(combined_tpm, aes(x = log10(nascent_tpm + 0.1), fill = rep_family))
g + geom_density(alpha = 0.4) + 
  ggtitle("Nascent-seq expression by repeat family")

g <- ggplot(combined_tpm, aes(x = rep_family, y = log10(nascent_tpm + 0.1)))
g + geom_boxplot() + 
  stat_compare_means() +
  ggtitle("Nascent-seq expression by repeat family")

g <- ggplot(combined_tpm, aes(x = fam, y = log10(nascent_tpm + 0.1)))
g + geom_boxplot() + 
  stat_compare_means() +
  ggtitle("Nascent-seq expression by repeat")
```









